// @commercetools-frontend/mc-scripts/dist/create-webpack-config-for-production-e57dacb7.esm.js
import _Object$keys from '@babel/runtime-corejs3/core-js-stable/object/keys';
import _Object$getOwnPropertySymbols from '@babel/runtime-corejs3/core-js-stable/object/get-own-property-symbols';
import _Object$getOwnPropertyDescriptor from '@babel/runtime-corejs3/core-js-stable/object/get-own-property-descriptor';
import _forEachInstanceProperty from '@babel/runtime-corejs3/core-js-stable/instance/for-each';
import _Object$getOwnPropertyDescriptors from '@babel/runtime-corejs3/core-js-stable/object/get-own-property-descriptors';
import _Object$defineProperties from '@babel/runtime-corejs3/core-js-stable/object/define-properties';
import _Object$defineProperty from '@babel/runtime-corejs3/core-js-stable/object/define-property';
import _defineProperty from '@babel/runtime-corejs3/helpers/esm/defineProperty';
import _filterInstanceProperty from '@babel/runtime-corejs3/core-js-stable/instance/filter';
import _mapInstanceProperty from '@babel/runtime-corejs3/core-js-stable/instance/map';
import _JSON$stringify from '@babel/runtime-corejs3/core-js-stable/json/stringify';
import _endsWithInstanceProperty from '@babel/runtime-corejs3/core-js-stable/instance/ends-with';
import _Number$isInteger from '@babel/runtime-corejs3/core-js-stable/number/is-integer';
import _concatInstanceProperty from '@babel/runtime-corejs3/core-js-stable/instance/concat';
import CssMinimizerPlugin from 'css-minimizer-webpack-plugin';
import HtmlWebpackPlugin from 'html-webpack-plugin';
import MiniCssExtractPlugin from 'mini-css-extract-plugin';
import MomentLocalesPlugin from 'moment-locales-webpack-plugin';
import TerserPlugin from 'terser-webpack-plugin';
import webpack from 'webpack';
import { BundleAnalyzerPlugin } from 'webpack-bundle-analyzer';
import _classCallCheck from '@babel/runtime-corejs3/helpers/esm/classCallCheck';
import _createClass from '@babel/runtime-corejs3/helpers/esm/createClass';
import _bindInstanceProperty from '@babel/runtime-corejs3/core-js-stable/instance/bind';
import _reduceInstanceProperty from '@babel/runtime-corejs3/core-js-stable/instance/reduce';
import _Object$assign from '@babel/runtime-corejs3/core-js-stable/object/assign';
import fs from 'fs';
import path from 'path';
import { c as createPostcssConfig } from './create-postcss-config-354ba07a.esm.js';
import { h as hasJsxRuntime } from './has-jsx-runtime-e3ecb09b.esm.js';
import { p as paths } from './paths-ded20ae1.esm.js';
import { v as vendorsToTranspile } from './vendors-to-transpile-7dc84a72.esm.js';

let FinalStatsWriterPlugin = /*#__PURE__*/function () {
  // Expected options:
  // - `outputPath`
  // - `includeFields`
  function FinalStatsWriterPlugin(config) {
    _classCallCheck(this, FinalStatsWriterPlugin);
    this.config = void 0;
    if (!config) throw new Error('Missing config options');
    this.config = config;
  }
  _createClass(FinalStatsWriterPlugin, [{
    key: "apply",
    value: function apply(compiler) {
      // This is the only hook that return the `stats` plugin
      // with the `time` info. It also contains all the stats that
      // we would get from the original `emit` hook.
      // https://webpack.js.org/api/compiler-hooks/#done
      // https://webpack.js.org/api/plugins/#plugin-types
      if (compiler.hooks) {
        var _context;
        compiler.hooks.done.tap('custom-stats-writer-plugin', _bindInstanceProperty(_context = this.writeStats).call(_context, this));
      } else {
        var _context2;
        // @ts-ignore
        compiler.plugin('done', _bindInstanceProperty(_context2 = this.writeStats).call(_context2, this));
      }
    }
  }, {
    key: "writeStats",
    value: function writeStats(stats) {
      let finalStats = stats.toJson();

      // Filter only included fields
      if (this.config.includeFields) {
        var _context3;
        finalStats = _reduceInstanceProperty(_context3 = this.config.includeFields).call(_context3, (aggregatedStats, key) =>
        // eslint-disable-next-line
        _Object$assign({}, aggregatedStats, {
          [key]: finalStats[key]
        }), {});
      }
      try {
        fs.accessSync(this.config.outputPath, fs.constants.F_OK);
        fs.writeFileSync(path.join(this.config.outputPath, 'stats.json'), _JSON$stringify(finalStats, null, 2));
      } catch (error) {
        // eslint-disable-next-line no-console
        console.warn("[FinalStatsWriterPlugin] The dist folder could not be found at ".concat(this.config.outputPath, ". Check the console for errors during the webpack compilation."));
      }
    }
  }]);
  return FinalStatsWriterPlugin;
}();

function ownKeys(e, r) { var t = _Object$keys(e); if (_Object$getOwnPropertySymbols) { var o = _Object$getOwnPropertySymbols(e); r && (o = _filterInstanceProperty(o).call(o, function (r) { return _Object$getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var _context7, _context8; var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? _forEachInstanceProperty(_context7 = ownKeys(Object(t), !0)).call(_context7, function (r) { _defineProperty(e, r, t[r]); }) : _Object$getOwnPropertyDescriptors ? _Object$defineProperties(e, _Object$getOwnPropertyDescriptors(t)) : _forEachInstanceProperty(_context8 = ownKeys(Object(t))).call(_context8, function (r) { _Object$defineProperty(e, r, _Object$getOwnPropertyDescriptor(t, r)); }); } return e; }
// https://babeljs.io/blog/2017/09/11/zero-config-with-babel-macros
const momentLocalesToKeep = ["de", "de-at", "de-ch", "en-au", "en-ca", "en-gb", "en-ie", "en-in", "en-nz", "en-sg", "es", "es-do", "es-mx", "fr", "pt-br", "zh-cn"];
const defaultToggleFlags = {
  // Allow to disable CSS extraction in case it's not necessary (e.g. for Storybook)
  enableExtractCss: true,
  // Allow to disable index.html generation in case it's not necessary (e.g. for Storybook)
  generateIndexHtml: true,
  // Some plugins spawn workers to speed up the build. However this can cause trouble on
  // certain machines local and CI. This flag set to limit or disable any parallelism.
  // Options:
  //    `true` to default to the machines number of CPUs
  //    `false` to disable any paralelism
  //    `int` for a specific number of CPUs
  parallelism: true,
  // Some environemnts do not require `core-js` and can hence disable
  // it explicitely. This will disable `core-js` for `preset-env` and the
  // `plugin-transform-runtime`.
  disableCoreJs: false
};
const defaultOptions = {
  entryPoint: paths.entryPoint,
  sourceFolders: paths.sourceFolders,
  postcssOptions: {},
  toggleFlags: defaultToggleFlags
};

/**
 * This is a factory function to create the default webpack config
 * for a MC Application in `production` mode.
 * The function requires the file path to the related application
 * "entry point".
 */
function createWebpackConfigForProduction() {
  var _context, _context2, _context3, _context4, _context5, _context6;
  let options = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  const mergedOptions = _objectSpread(_objectSpread(_objectSpread({}, defaultOptions), options), {}, {
    toggleFlags: _objectSpread(_objectSpread({}, defaultToggleFlags), options.toggleFlags)
  });
  return {
    // Don't attempt to continue if there are any errors.
    bail: true,
    // https://webpack.js.org/concepts/#mode
    mode: 'production',
    // We generate sourcemaps in production. This is slow but gives good results.
    // Sourcemaps are pushed to Google Storage and Sentry.
    // https://webpack.js.org/configuration/devtool/#devtool
    devtool: 'source-map',
    // https://medium.com/webpack/webpack-4-code-splitting-chunk-graph-and-the-splitchunks-optimization-be739a861366
    // https://medium.com/webpack/webpack-4-mode-and-optimization-5423a6bc597a
    optimization: {
      minimizer: _filterInstanceProperty(_context = [new TerserPlugin({
        terserOptions: {
          parse: {
            // we want terser to parse ecma 8 code. However, we don't want it
            // to apply any minification steps that turns valid ecma 5 code
            // into invalid ecma 5 code. This is why the 'compress' and 'output'
            // sections only apply transformations that are ecma 5 safe
            // https://github.com/facebook/create-react-app/pull/4234
            ecma: 2018
          },
          compress: {
            ecma: 5,
            // Disabled because of an issue with Uglify breaking seemingly valid code:
            // https://github.com/facebook/create-react-app/issues/2376
            // Pending further investigation:
            // https://github.com/mishoo/UglifyJS2/issues/2011
            comparisons: false,
            // Disabled because of an issue with Terser breaking valid code:
            // https://github.com/facebook/create-react-app/issues/5250
            // Pending futher investigation:
            // https://github.com/terser-js/terser/issues/120
            inline: 2
          },
          mangle: {
            safari10: true
          },
          // Added for profiling in devtools
          keep_classnames: true,
          keep_fnames: true,
          output: {
            ecma: 5,
            comments: false,
            // Turned on because emoji and regex is not minified properly using default
            // https://github.com/facebook/create-react-app/issues/2488
            ascii_only: true
          }
        },
        // Use multi-process parallel running to improve the build speed
        // Default number of concurrent runs: os.cpus().length - 1
        parallel: mergedOptions.toggleFlags.parallelism
      }), mergedOptions.toggleFlags.enableExtractCss && new CssMinimizerPlugin()]).call(_context, Boolean),
      // Keep the runtime chunk separated to enable long term caching
      runtimeChunk: {
        name: 'runtime'
      },
      moduleIds: 'named',
      chunkIds: 'deterministic'
    },
    resolve: {
      // These are the reasonable defaults supported by the Node ecosystem.
      // We also include JSX as a common component filename extension to support
      // some tools, although we do not recommend using it, see:
      // https://github.com/facebook/create-react-app/issues/290
      // `web` extension prefixes have been added for better support
      // for React Native Web.
      extensions: _mapInstanceProperty(_context2 = ['js', 'mjs', 'cjs', 'ts', 'tsx', 'json', 'jsx']).call(_context2, ext => ".".concat(ext)),
      // NOTE: this is meant to be a temporary list of fallback/polyfills for certain
      // nodejs modules. With Webpack <5 these polyfills were included by default in Webpack,
      // however now it's not the case anymore.
      // See also related work in CRA: https://github.com/facebook/create-react-app/pull/11764
      fallback: {
        querystring: require.resolve('querystring-es3')
      }
    },
    // In production, we only want to load the polyfills and the app code.
    entry: {
      app: _filterInstanceProperty(_context3 = [require.resolve('@commercetools-frontend/mc-scripts/application-runtime'), !mergedOptions.toggleFlags.disableCoreJs && require.resolve('core-js/stable'), mergedOptions.entryPoint]).call(_context3, Boolean)
    },
    output: {
      // Generated JS file names (with nested folders).
      // There will be one main bundle, and one file per asynchronous chunk.
      filename: '[name].[chunkhash].js',
      chunkFilename: '[id].[name].[chunkhash].js',
      // The build folder.
      path: paths.appBuild,
      pathinfo: false,
      // Will be injected on runtime. See `packages/application-shell/src/public-path.js`
      publicPath: ''
    },
    plugins: _filterInstanceProperty(_context4 = [
    // Allows to "assign" custom options to the `webpack` object.
    // At the moment, this is used to share some props with `postcss.config`.
    new webpack.LoaderOptionsPlugin({
      options: {
        sourceFolders: mergedOptions.sourceFolders,
        context: __dirname
      }
    }),
    // Makes some environment variables available to the JS code, for example:
    // if (process.env.NODE_ENV === 'production') { ... }.
    new webpack.DefinePlugin({
      __DEV__: 'false',
      'process.env': {
        NODE_ENV: _JSON$stringify('production')
      }
    }),
    // Only keep locales that are available in the Merchant Center.
    new MomentLocalesPlugin({
      localesToKeep: momentLocalesToKeep
    }),
    // Generate a `stats.json` file containing information and paths to
    // the assets that webpack created.
    // This is necessary to programmatically refer to the correct bundle path
    // in the `index.html`.
    new FinalStatsWriterPlugin({
      outputPath: paths.appBuild,
      includeFields: ['entrypoints', 'assets', 'publicPath', 'time']
    }), mergedOptions.toggleFlags.generateIndexHtml && new HtmlWebpackPlugin({
      inject: false,
      filename: 'index.html.template',
      template: require.resolve('@commercetools-frontend/mc-html-template/webpack')
    }), mergedOptions.toggleFlags.enableExtractCss &&
    // Extracts CSS into one CSS file to mimic CSS order in dev
    new MiniCssExtractPlugin({
      filename: '[name].[chunkhash].css',
      chunkFilename: '[id].[name].[chunkhash].css'
    }), process.env.ANALYZE_BUNDLE === 'true' && new BundleAnalyzerPlugin({
      defaultSizes: 'gzip'
    })]).call(_context4, Boolean),
    module: {
      // Makes missing exports an error instead of warning.
      strictExportPresence: true,
      rules: _filterInstanceProperty(_context5 = [
      // For svg icons, we want to get them transformed into React components
      // when we import them.
      {
        test: /\.react\.svg$/,
        use: [{
          loader: require.resolve('babel-loader'),
          options: {
            babelrc: false,
            presets: [[require.resolve('@commercetools-frontend/babel-preset-mc-app'), {
              runtime: hasJsxRuntime() ? 'automatic' : 'classic',
              disableCoreJs: mergedOptions.toggleFlags.disableCoreJs
            }]],
            // This is a feature of `babel-loader` for webpack (not Babel itself).
            // It enables caching results in ./node_modules/.cache/babel-loader/
            // directory for faster rebuilds.
            cacheDirectory: true,
            highlightCode: true
          }
        }, {
          loader: require.resolve('@svgr/webpack'),
          options: {
            icon: false,
            svgoConfig: {
              plugins: [{
                // https://github.com/svg/svgo#default-preset
                name: 'preset-default',
                params: {
                  overrides: {
                    removeViewBox: false
                  }
                }
              }]
            }
          }
        }]
      },
      // For normal svg files (not icons) we should load the file normally
      // and simply use it as a `<img src/>`.
      {
        test: function testForNormalSvgFiles(fileName) {
          return (
            // Use this only for plain SVG.
            // For SVG as React components, see loader above.
            _endsWithInstanceProperty(fileName).call(fileName, '.svg') && !_endsWithInstanceProperty(fileName).call(fileName, '.react.svg')
          );
        },
        use: [{
          loader: require.resolve('svg-url-loader')
        }]
      }, {
        test: /\.png$/,
        type: 'asset/resource'
      },
      // "postcss" loader applies autoprefixer to our CSS
      // "css" loader resolves paths in CSS and adds assets as dependencies.
      // "style" loader turns CSS into JS modules that inject <style> tags.
      // In production, we use MiniCSSExtractPlugin to extract that CSS
      // to a file, but in development "style" loader enables hot editing
      // of CSS.
      // By default we support CSS Modules with the extension `.mod.css` and `.module.css`.
      {
        test: /\.(mod|module)\.css$/,
        include: mergedOptions.sourceFolders,
        use: [mergedOptions.toggleFlags.enableExtractCss ? MiniCssExtractPlugin.loader : require.resolve('style-loader'), {
          loader: require.resolve('css-loader'),
          options: {
            modules: {
              mode: 'local',
              localIdentName: '[name]__[local]___[hash:base64:5]',
              localIdentHashSalt: 'ct'
            },
            importLoaders: 1
          }
        }, {
          loader: require.resolve('postcss-loader'),
          options: {
            postcssOptions: createPostcssConfig(mergedOptions.postcssOptions)
          }
        }]
      }, {
        test: function testForNormalCssFiles(fileName) {
          return (
            // Use this only for plain CSS.
            // For css-modules, see loader above.
            _endsWithInstanceProperty(fileName).call(fileName, '.css') && !(_endsWithInstanceProperty(fileName).call(fileName, '.mod.css') || _endsWithInstanceProperty(fileName).call(fileName, '.module.css'))
          );
        },
        // "postcss" loader applies autoprefixer to our CSS.
        // "css" loader resolves paths in CSS and adds assets as dependencies.
        // "MiniCssExtractPlugin" or "style" loader extracts css to one file per css file.
        oneOf: [{
          // Use "postcss" for all the included source folders.
          include: mergedOptions.sourceFolders,
          use: [mergedOptions.toggleFlags.enableExtractCss ? MiniCssExtractPlugin.loader : require.resolve('style-loader'), {
            loader: require.resolve('css-loader'),
            options: {
              modules: {
                mode: 'icss'
              },
              importLoaders: 1
            }
          }, {
            loader: require.resolve('postcss-loader'),
            options: {
              postcssOptions: createPostcssConfig(mergedOptions.postcssOptions)
            }
          }]
        }, {
          // For all other vendor CSS, do not use "postcss" loader.
          // But still use MiniCssExtractPlugin :)
          include: /node_modules/,
          use: [mergedOptions.toggleFlags.enableExtractCss ? MiniCssExtractPlugin.loader : require.resolve('style-loader'), {
            loader: require.resolve('css-loader'),
            options: {
              modules: {
                mode: 'icss'
              },
              importLoaders: 1
            }
          }],
          // Don't consider CSS imports dead code even if the
          // containing package claims to have no side effects.
          // Remove this when webpack adds a warning or an error for this.
          // See https://github.com/webpack/webpack/issues/6571
          sideEffects: true
        }]
      },
      // Fix for react-intl
      // https://github.com/formatjs/formatjs/issues/143#issuecomment-518774786
      {
        test: /\.mjs$/,
        type: 'javascript/auto',
        resolve: {
          // https://webpack.js.org/configuration/module/#resolvefullyspecified
          fullySpecified: false
        }
      },
      // Process application JavaScript with Babel.
      {
        test: /\.(js|mjs|cjs|jsx|ts|tsx)$/,
        use: [
        // This loader parallelizes code compilation, it is optional but
        // improves compile time on larger projects
        {
          loader: require.resolve('thread-loader'),
          options: _objectSpread({}, _Number$isInteger(mergedOptions.toggleFlags.parallelism) ? {
            workers: mergedOptions.toggleFlags.parallelism
          } : {})
        }, {
          loader: require.resolve('babel-loader'),
          options: {
            babelrc: false,
            configFile: false,
            compact: false,
            presets: [[require.resolve('@commercetools-frontend/babel-preset-mc-app'), {
              runtime: hasJsxRuntime() ? 'automatic' : 'classic'
            }]],
            // This is a feature of `babel-loader` for webpack (not Babel itself).
            // It enables caching results in ./node_modules/.cache/babel-loader/
            // directory for faster rebuilds.
            cacheDirectory: true,
            highlightCode: true
          }
        }],
        include: _concatInstanceProperty(_context6 = mergedOptions.sourceFolders).call(_context6, vendorsToTranspile),
        // Disable require.ensure as it's not a standard language feature.
        parser: {
          requireEnsure: false
        }
      },
      // Allow to import `*.graphql` SDL files.
      {
        test: /\.graphql$/,
        include: mergedOptions.sourceFolders,
        exclude: /node_modules/,
        use: [require.resolve('graphql-tag/loader')]
      }]).call(_context5, Boolean)
    },
    // Turn off performance processing because we utilize
    // our own hints via the FileSizeReporter
    performance: false
  };
}

export { createWebpackConfigForProduction as c };
