// @commercetools-frontend/mc-scripts/dist/vite-plugin-svgr-5c953fe7.esm.js
import fs from 'fs';
import { createFilter } from '@rollup/pluginutils';
import { transformWithEsbuild } from 'vite';

/**
 * COPIED FROM https://github.com/pd4d10/vite-plugin-svgr
 */
function vitePluginSvgr() {
  const filter = createFilter('**/*.react.svg');
  return {
    name: 'vite-plugin-svgr',
    async transform(_code, id) {
      if (filter(id)) {
        const _await$import = await import('@svgr/core'),
          transform = _await$import.transform;
        const svgCode = await fs.promises.readFile(id, 'utf8');
        const componentCode = await transform(svgCode, {
          icon: false,
          svgoConfig: {
            plugins: [{
              // https://github.com/svg/svgo#default-preset
              name: 'preset-default',
              params: {
                overrides: {
                  removeViewBox: false
                }
              }
            }]
          }
        }, {
          filePath: id,
          caller: {
            previousExport: null
          }
        });
        const res = await transformWithEsbuild(componentCode, id, {
          loader: 'jsx'
        });
        return {
          code: res.code,
          map: null // TODO:
        };
      }

      return null;
    }
  };
}

export { vitePluginSvgr as v };
