// @commercetools-frontend/mc-scripts/dist/vite-plugin-svgr-a732f8df.cjs.prod.js
'use strict';

var fs = require('fs');
var pluginutils = require('@rollup/pluginutils');
var vite = require('vite');

function _interopDefault (e) { return e && e.__esModule ? e : { 'default': e }; }

function _interopNamespace(e) {
  if (e && e.__esModule) return e;
  var n = Object.create(null);
  if (e) {
    Object.keys(e).forEach(function (k) {
      if (k !== 'default') {
        var d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: function () { return e[k]; }
        });
      }
    });
  }
  n["default"] = e;
  return Object.freeze(n);
}

var fs__default = /*#__PURE__*/_interopDefault(fs);

/**
 * COPIED FROM https://github.com/pd4d10/vite-plugin-svgr
 */
function vitePluginSvgr() {
  const filter = pluginutils.createFilter('**/*.react.svg');
  return {
    name: 'vite-plugin-svgr',
    async transform(_code, id) {
      if (filter(id)) {
        const _await$import = await Promise.resolve().then(function () { return /*#__PURE__*/_interopNamespace(require('@svgr/core')); }),
          transform = _await$import.transform;
        const svgCode = await fs__default["default"].promises.readFile(id, 'utf8');
        const componentCode = await transform(svgCode, {
          icon: false,
          svgoConfig: {
            plugins: [{
              // https://github.com/svg/svgo#default-preset
              name: 'preset-default',
              params: {
                overrides: {
                  removeViewBox: false
                }
              }
            }]
          }
        }, {
          filePath: id,
          caller: {
            previousExport: null
          }
        });
        const res = await vite.transformWithEsbuild(componentCode, id, {
          loader: 'jsx'
        });
        return {
          code: res.code,
          map: null // TODO:
        };
      }

      return null;
    }
  };
}

exports.vitePluginSvgr = vitePluginSvgr;
