// @commercetools-uikit/buttons/dist/commercetools-uikit-buttons.cjs.prod.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var accessibleButton = require('@commercetools-uikit/accessible-button');
var flatButton = require('@commercetools-uikit/flat-button');
var primaryButton = require('@commercetools-uikit/primary-button');
var iconButton = require('@commercetools-uikit/icon-button');
var linkButton = require('@commercetools-uikit/link-button');
var secondaryButton = require('@commercetools-uikit/secondary-button');
var secondaryIconButton = require('@commercetools-uikit/secondary-icon-button');

function _interopDefault (e) { return e && e.__esModule ? e : { 'default': e }; }

var accessibleButton__default = /*#__PURE__*/_interopDefault(accessibleButton);
var flatButton__default = /*#__PURE__*/_interopDefault(flatButton);
var primaryButton__default = /*#__PURE__*/_interopDefault(primaryButton);
var iconButton__default = /*#__PURE__*/_interopDefault(iconButton);
var linkButton__default = /*#__PURE__*/_interopDefault(linkButton);
var secondaryButton__default = /*#__PURE__*/_interopDefault(secondaryButton);
var secondaryIconButton__default = /*#__PURE__*/_interopDefault(secondaryIconButton);



Object.defineProperty(exports, 'AccessibleButton', {
	enumerable: true,
	get: function () { return accessibleButton__default["default"]; }
});
Object.defineProperty(exports, 'FlatButton', {
	enumerable: true,
	get: function () { return flatButton__default["default"]; }
});
Object.defineProperty(exports, 'PrimaryButton', {
	enumerable: true,
	get: function () { return primaryButton__default["default"]; }
});
Object.defineProperty(exports, 'IconButton', {
	enumerable: true,
	get: function () { return iconButton__default["default"]; }
});
Object.defineProperty(exports, 'LinkButton', {
	enumerable: true,
	get: function () { return linkButton__default["default"]; }
});
Object.defineProperty(exports, 'SecondaryButton', {
	enumerable: true,
	get: function () { return secondaryButton__default["default"]; }
});
Object.defineProperty(exports, 'SecondaryIconButton', {
	enumerable: true,
	get: function () { return secondaryIconButton__default["default"]; }
});
