// @commercetools-uikit/async-select-input/dist/commercetools-uikit-async-select-input.esm.js
import _Object$keys from '@babel/runtime-corejs3/core-js-stable/object/keys';
import _Object$getOwnPropertySymbols from '@babel/runtime-corejs3/core-js-stable/object/get-own-property-symbols';
import _filterInstanceProperty from '@babel/runtime-corejs3/core-js-stable/instance/filter';
import _Object$getOwnPropertyDescriptor from '@babel/runtime-corejs3/core-js-stable/object/get-own-property-descriptor';
import _forEachInstanceProperty from '@babel/runtime-corejs3/core-js-stable/instance/for-each';
import _Object$getOwnPropertyDescriptors from '@babel/runtime-corejs3/core-js-stable/object/get-own-property-descriptors';
import _Object$defineProperties from '@babel/runtime-corejs3/core-js-stable/object/define-properties';
import _Object$defineProperty from '@babel/runtime-corejs3/core-js-stable/object/define-property';
import _defineProperty from '@babel/runtime-corejs3/helpers/esm/defineProperty';
import _pt from 'prop-types';
import { useIntl } from 'react-intl';
import isEmpty from 'lodash/isEmpty';
import { components } from 'react-select';
import AsyncSelect from 'react-select/async';
import { warning, filterDataAttributes } from '@commercetools-uikit/utils';
import Constraints from '@commercetools-uikit/constraints';
import LoadingSpinner from '@commercetools-uikit/loading-spinner';
import { warnIfMenuPortalPropsAreMissing, messages, customComponentsWithIcons, createSelectStyles, DropdownIndicator, ClearIndicator, TagRemove } from '@commercetools-uikit/select-utils';
import { jsx } from '@emotion/react/jsx-runtime';

function ownKeys(object, enumerableOnly) { var keys = _Object$keys(object); if (_Object$getOwnPropertySymbols) { var symbols = _Object$getOwnPropertySymbols(object); enumerableOnly && (symbols = _filterInstanceProperty(symbols).call(symbols, function (sym) { return _Object$getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }
function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var _context, _context2; var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? _forEachInstanceProperty(_context = ownKeys(Object(source), !0)).call(_context, function (key) { _defineProperty(target, key, source[key]); }) : _Object$getOwnPropertyDescriptors ? _Object$defineProperties(target, _Object$getOwnPropertyDescriptors(source)) : _forEachInstanceProperty(_context2 = ownKeys(Object(source))).call(_context2, function (key) { _Object$defineProperty(target, key, _Object$getOwnPropertyDescriptor(source, key)); }); } return target; }
const LoadingIndicator = () => jsx(LoadingSpinner, {
  scale: "s"
});
LoadingIndicator.displayName = 'LoadingIndicator';
const customizedComponents = {
  DropdownIndicator,
  ClearIndicator,
  LoadingIndicator,
  MultiValueRemove: TagRemove
};
const defaultProps = {
  // Using "null" will ensure that the currently selected value disappears in
  // case "undefined" gets passed as the next value
  value: null,
  isSearchable: true,
  menuPortalZIndex: 1,
  controlShouldRenderValue: true
};
const AsyncSelectInput = props => {
  const intl = useIntl();
  if (!props.isReadOnly) {
    process.env.NODE_ENV !== "production" ? warning(typeof props.onChange === 'function', 'AsyncSelectInput: `onChange` is required when input is not read only.') : void 0;
  }
  warnIfMenuPortalPropsAreMissing({
    menuPortalZIndex: props.menuPortalZIndex,
    menuPortalTarget: props.menuPortalTarget,
    componentName: 'AsyncSelectInput'
  });
  const placeholder = props.placeholder || intl.formatMessage(messages.placeholder);
  const loadingMessage = () => {
    if (typeof props.loadingMessage === 'function') {
      return props.loadingMessage();
    }
    return props.loadingMessage || intl.formatMessage(messages.loadingOptions);
  };
  return jsx(Constraints.Horizontal, {
    max: props.horizontalConstraint,
    children: jsx("div", _objectSpread(_objectSpread({}, filterDataAttributes(props)), {}, {
      children: jsx(AsyncSelect, {
        "aria-label": props['aria-label'],
        "aria-labelledby": props['aria-labelledby'],
        "aria-invalid": props['aria-invalid'],
        "aria-errormessage": props['aria-errormessage'],
        autoFocus: props.isAutofocussed,
        backspaceRemovesValue: props.isReadOnly ? false : props.backspaceRemovesValue,
        components: _objectSpread(_objectSpread(_objectSpread(_objectSpread({}, customizedComponents), props.iconLeft && !props.isMulti ? customComponentsWithIcons : {}), props.isReadOnly ? {
          Input: ownProps => jsx(components.Input, _objectSpread(_objectSpread({}, ownProps), {}, {
            readOnly: true
          }))
        } : {}), props.components),
        menuIsOpen: props.isReadOnly ? false : props.menuIsOpen,
        styles: createSelectStyles({
          hasWarning: props.hasWarning,
          hasError: props.hasError,
          showOptionGroupDivider: props.showOptionGroupDivider,
          menuPortalZIndex: props.menuPortalZIndex,
          isDisabled: props.isDisabled,
          isReadOnly: props.isReadOnly,
          iconLeft: props.iconLeft,
          isMulti: props.isMulti,
          hasValue: !isEmpty(props.value),
          controlShouldRenderValue: props.controlShouldRenderValue,
          horizontalConstraint: props.horizontalConstraint
        }),
        filterOption: props.filterOption
        // react-select uses "id" (for the container) and "inputId" (for the input),
        // but we use "id" (for the input) and "containerId" (for the container)
        // instead.
        // This makes it easier to less confusing to use with <label />s.
        ,
        id: props.containerId,
        inputId: props.id,
        inputValue: props.inputValue,
        isClearable: props.isReadOnly ? false : props.isClearable,
        isDisabled: props.isDisabled,
        isOptionDisabled: props.isOptionDisabled,
        isMulti: props.isMulti,
        isSearchable: props.isSearchable,
        maxMenuHeight: props.maxMenuHeight,
        menuPortalTarget: props.menuPortalTarget,
        menuShouldBlockScroll: props.menuShouldBlockScroll,
        closeMenuOnSelect: props.closeMenuOnSelect,
        name: props.name,
        loadingMessage: loadingMessage,
        noOptionsMessage: props.noOptionsMessage || (_ref => {
          let inputValue = _ref.inputValue;
          return inputValue === '' ? intl.formatMessage(messages.noOptionsMessageWithoutInputValue) : intl.formatMessage(messages.noOptionsMessageWithInputValue, {
            inputValue
          });
        }),
        onBlur: props.onBlur ? () => {
          const event = {
            target: {
              id: props.id,
              name: (() => {
                if (!props.name) return undefined;
                if (!props.isMulti) return props.name;
                // We append the ".0" to make Formik set the touched
                // state as an array instead of a boolean only.
                // Otherwise the shapes would clash on submission, as
                // Formik will create an array on submission anyways.
                return "".concat(props.name, ".0");
              })()
            },
            persist: () => {}
          };
          props.onBlur && props.onBlur(event);
        } : undefined,
        onChange: (value, info) => {
          var _props$onChange;
          let newValue = value;
          if (props.isMulti && !newValue) {
            newValue = [];
          }
          (_props$onChange = props.onChange) === null || _props$onChange === void 0 ? void 0 : _props$onChange.call(props, {
            target: {
              id: props.id,
              name: props.name,
              value: newValue
            },
            persist: () => {}
          }, info);
        },
        onFocus: props.onFocus,
        onInputChange: props.onInputChange,
        placeholder: placeholder,
        tabIndex: props.tabIndex,
        tabSelectsValue: props.tabSelectsValue,
        value: props.value
        // Async react-select props
        ,
        defaultOptions: props.defaultOptions,
        loadOptions: props.loadOptions,
        cacheOptions: props.cacheOptions
        // Extra props
        // @ts-ignore: passed to the react-select components via `selectProps`.
        ,
        iconLeft: props.iconLeft,
        controlShouldRenderValue: props.controlShouldRenderValue
      })
    }))
  });
};
AsyncSelectInput.propTypes = process.env.NODE_ENV !== "production" ? {
  horizontalConstraint: _pt.oneOf([3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 'scale', 'auto']),
  hasError: _pt.bool,
  hasWarning: _pt.bool,
  isReadOnly: _pt.bool,
  isAutofocussed: _pt.bool,
  menuPortalZIndex: _pt.number.isRequired,
  onBlur: _pt.func,
  onChange: _pt.func,
  loadingMessage: _pt.oneOfType([_pt.string, _pt.func]),
  defaultOptions: _pt.oneOfType([_pt.any, _pt.bool]),
  showOptionGroupDivider: _pt.bool,
  iconLeft: _pt.node
} : {};
AsyncSelectInput.displayName = 'AsyncSelectInput';
AsyncSelectInput.defaultProps = defaultProps;

/**
 * Expose static helper methods.
 */

// Formik will set the field to an array on submission, so we always have to
// deal with an array. The touched state ends up being an empty array in case
// values were removed only. So we have to treat any array we receive as
// a signal of the field having been touched.
AsyncSelectInput.isTouched = touched => Boolean(touched);

/**
 * Expose react-select components for customization purposes.
 */

// custom
AsyncSelectInput.ClearIndicator = customizedComponents.ClearIndicator;
AsyncSelectInput.Control = components.Control;
AsyncSelectInput.CrossIcon = components.CrossIcon;
AsyncSelectInput.DownChevron = components.DownChevron;
// custom
AsyncSelectInput.DropdownIndicator = customizedComponents.DropdownIndicator;
AsyncSelectInput.Group = components.Group;
AsyncSelectInput.GroupHeading = components.GroupHeading;
AsyncSelectInput.IndicatorSeparator = components.IndicatorSeparator;
AsyncSelectInput.IndicatorsContainer = components.IndicatorsContainer;
AsyncSelectInput.Input = components.Input;
// custom
AsyncSelectInput.LoadingIndicator = customizedComponents.LoadingIndicator;
AsyncSelectInput.LoadingMessage = components.LoadingMessage;
AsyncSelectInput.Menu = components.Menu;
AsyncSelectInput.MenuList = components.MenuList;
AsyncSelectInput.MenuPortal = components.MenuPortal;
AsyncSelectInput.MultiValue = components.MultiValue;
AsyncSelectInput.MultiValueContainer = components.MultiValueContainer;
AsyncSelectInput.MultiValueLabel = components.MultiValueLabel;
// custom
AsyncSelectInput.MultiValueRemove = customizedComponents.MultiValueRemove;
AsyncSelectInput.NoOptionsMessage = components.NoOptionsMessage;
AsyncSelectInput.Option = components.Option;
AsyncSelectInput.Placeholder = components.Placeholder;
AsyncSelectInput.SelectContainer = components.SelectContainer;
AsyncSelectInput.SingleValue = components.SingleValue;
AsyncSelectInput.ValueContainer = components.ValueContainer;
var AsyncSelectInput$1 = AsyncSelectInput;

// NOTE: This string will be replaced on build time with the package version.
var version = "16.11.0";

export { AsyncSelectInput$1 as default, version };
