// @commercetools-frontend/application-components/dist/public-page-layout-295bf2b5.cjs.prod.js
'use strict';

var _Object$keys = require('@babel/runtime-corejs3/core-js-stable/object/keys');
var _Object$getOwnPropertySymbols = require('@babel/runtime-corejs3/core-js-stable/object/get-own-property-symbols');
var _filterInstanceProperty = require('@babel/runtime-corejs3/core-js-stable/instance/filter');
var _Object$getOwnPropertyDescriptor = require('@babel/runtime-corejs3/core-js-stable/object/get-own-property-descriptor');
var _forEachInstanceProperty = require('@babel/runtime-corejs3/core-js-stable/instance/for-each');
var _Object$getOwnPropertyDescriptors = require('@babel/runtime-corejs3/core-js-stable/object/get-own-property-descriptors');
var _Object$defineProperties = require('@babel/runtime-corejs3/core-js-stable/object/define-properties');
var _Object$defineProperty = require('@babel/runtime-corejs3/core-js-stable/object/define-property');
var _defineProperty = require('@babel/runtime-corejs3/helpers/defineProperty');
var _styled = require('@emotion/styled/base');
require('prop-types');
require('react');
var react = require('@emotion/react');
var CommercetoolsLogoOnWhiteSvg = require('@commercetools-frontend/assets/logos/color-on-white-horizontal.svg');
var designSystem = require('@commercetools-uikit/design-system');
var Spacings = require('@commercetools-uikit/spacings');
var Text = require('@commercetools-uikit/text');
var dist_commercetoolsFrontendApplicationComponents = require('./commercetools-frontend-application-components.cjs.prod.js');
var jsxRuntime = require('@emotion/react/jsx-runtime');
require('react-intl');
require('react-router-dom');
require('@commercetools-uikit/utils');
require('@babel/runtime-corejs3/core-js-stable/instance/concat');
require('react-modal');
require('@commercetools-frontend/constants');
require('@commercetools-uikit/card');
require('@commercetools-uikit/icons');
require('@commercetools-uikit/secondary-icon-button');
require('@commercetools-frontend/i18n');
require('@commercetools-uikit/primary-button');
require('@commercetools-uikit/secondary-button');
require('@babel/runtime-corejs3/core-js-stable/instance/starts-with');
require('lodash/omitBy');
require('@babel/runtime-corejs3/helpers/slicedToArray');
require('@babel/runtime-corejs3/core-js-stable/set-timeout');
require('@babel/runtime-corejs3/core-js-stable/instance/map');
require('@commercetools-frontend/sentry');
require('@commercetools-uikit/constraints');
require('@commercetools-frontend/actions-global');
require('@commercetools-frontend/application-shell-connectors');
require('@commercetools-uikit/icon-button');
require('@commercetools-uikit/flat-button');
require('@flopflip/react-broadcast');
require('@babel/runtime-corejs3/core-js-stable/instance/find');
require('@babel/runtime-corejs3/core-js-stable/object/entries');
require('@commercetools-frontend/assets/images/page-not-found.svg');
require('@commercetools-frontend/assets/images/doors-closed.svg');
require('@react-hook/resize-observer');
require('@commercetools-uikit/hooks');

function _interopDefault (e) { return e && e.__esModule ? e : { 'default': e }; }

var _Object$keys__default = /*#__PURE__*/_interopDefault(_Object$keys);
var _Object$getOwnPropertySymbols__default = /*#__PURE__*/_interopDefault(_Object$getOwnPropertySymbols);
var _filterInstanceProperty__default = /*#__PURE__*/_interopDefault(_filterInstanceProperty);
var _Object$getOwnPropertyDescriptor__default = /*#__PURE__*/_interopDefault(_Object$getOwnPropertyDescriptor);
var _forEachInstanceProperty__default = /*#__PURE__*/_interopDefault(_forEachInstanceProperty);
var _Object$getOwnPropertyDescriptors__default = /*#__PURE__*/_interopDefault(_Object$getOwnPropertyDescriptors);
var _Object$defineProperties__default = /*#__PURE__*/_interopDefault(_Object$defineProperties);
var _Object$defineProperty__default = /*#__PURE__*/_interopDefault(_Object$defineProperty);
var _styled__default = /*#__PURE__*/_interopDefault(_styled);
var CommercetoolsLogoOnWhiteSvg__default = /*#__PURE__*/_interopDefault(CommercetoolsLogoOnWhiteSvg);
var Spacings__default = /*#__PURE__*/_interopDefault(Spacings);
var Text__default = /*#__PURE__*/_interopDefault(Text);

function ownKeys(e, r) { var t = _Object$keys__default["default"](e); if (_Object$getOwnPropertySymbols__default["default"]) { var o = _Object$getOwnPropertySymbols__default["default"](e); r && (o = _filterInstanceProperty__default["default"](o).call(o, function (r) { return _Object$getOwnPropertyDescriptor__default["default"](e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var _context, _context2; var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? _forEachInstanceProperty__default["default"](_context = ownKeys(Object(t), !0)).call(_context, function (r) { _defineProperty(e, r, t[r]); }) : _Object$getOwnPropertyDescriptors__default["default"] ? _Object$defineProperties__default["default"](e, _Object$getOwnPropertyDescriptors__default["default"](t)) : _forEachInstanceProperty__default["default"](_context2 = ownKeys(Object(t))).call(_context2, function (r) { _Object$defineProperty__default["default"](e, r, _Object$getOwnPropertyDescriptor__default["default"](t, r)); }); } return e; }
const year = new Date().getUTCFullYear();
const Container = /*#__PURE__*/_styled__default["default"]("div", {
  target: "e6h44oa2"
} )("width:100%;min-height:100vh;display:flex;padding:", designSystem.customProperties.spacingXl, " 0;justify-content:center;background-size:cover;background-position:center;" + ("" ));
const ContainerColumn = /*#__PURE__*/_styled__default["default"]("div", {
  target: "e6h44oa1"
} )("width:", dist_commercetoolsFrontendApplicationComponents.designTokens.widthForPageLayoutContentColumn, ";" + ("" ));
const ContainerColumnWide = /*#__PURE__*/_styled__default["default"]("div", {
  target: "e6h44oa0"
} )("width:", designSystem.customProperties.constraint15, ";" + ("" ));
const PublicPageLayoutContent = props => {
  if (props.contentScale === 'wide') {
    return jsxRuntime.jsx(ContainerColumnWide, {
      children: props.children
    });
  }
  return jsxRuntime.jsx(ContainerColumn, {
    children: props.children
  });
};
PublicPageLayoutContent.propTypes = {};
const PublicPageLayout = props => {
  return jsxRuntime.jsx(Container, {
    children: jsxRuntime.jsxs(Spacings__default["default"].Stack, {
      scale: "xl",
      alignItems: "center",
      children: [jsxRuntime.jsx(ContainerColumn, {
        children: jsxRuntime.jsx("div", {
          children: jsxRuntime.jsx("img", {
            src: CommercetoolsLogoOnWhiteSvg__default["default"],
            alt: "commercetools logo"
          })
        })
      }), props.welcomeMessage && jsxRuntime.jsx(ContainerColumn, {
        children: jsxRuntime.jsx(Text__default["default"].Headline, {
          as: "h2",
          children: jsxRuntime.jsx("div", {
            css: /*#__PURE__*/react.css("color:", designSystem.customProperties.colorSolid, ";text-align:center;" + ("" ), "" ),
            children: props.welcomeMessage
          })
        })
      }), jsxRuntime.jsxs(Spacings__default["default"].Stack, {
        scale: "xl",
        children: [jsxRuntime.jsx(PublicPageLayoutContent, _objectSpread({}, props)), jsxRuntime.jsx(PublicPageLayoutContent, {
          contentScale: props.contentScale,
          children: jsxRuntime.jsxs(Spacings__default["default"].Stack, {
            scale: "xs",
            alignItems: props.contentScale === 'wide' ? 'center' : 'stretch',
            children: [props.legalMessage && jsxRuntime.jsx(Text__default["default"].Detail, {
              tone: "secondary",
              children: props.legalMessage
            }), jsxRuntime.jsx(Text__default["default"].Detail, {
              tone: "secondary",
              children: "".concat(year, " \xA9 commercetools")
            })]
          })
        })]
      })]
    })
  });
};
PublicPageLayout.propTypes = {};
PublicPageLayout.displayName = 'PublicPageLayout';
PublicPageLayout.defaultProps = {
  contentScale: 'normal'
};

exports["default"] = PublicPageLayout;
