// @commercetools-frontend/application-shell/dist/commercetools-frontend-application-shell.esm.js
export { g as ApplicationPageTitle, A as ApplicationShell, f as ApplicationShellProvider, C as ConfigureIntlProvider, h as CustomViewShell, I as InjectReducers, S as SetupFlopFlipProvider, i as SuspendedRoute, s as setupGlobalErrorListener, u as useRoutesCreator, d as version } from './index-2b55a829.esm.js';
export { R as RouteCatchAll, a as applyDefaultMiddlewares } from './index-89e5711a.esm.js';
export { buildApiUrl, createApolloClient, createApolloContextForProxyForwardTo, createHttpClientOptions, executeHttpClientRequest, getMcApiUrl, selectProjectKeyFromUrl, selectUserId, useMcLazyQuery, useMcMutation, useMcQuery } from '@commercetools-frontend/application-shell-connectors';
export { entryPointUriPathToPermissionKeys, entryPointUriPathToResourceAccesses } from '@commercetools-frontend/application-config/ssr';
export { ConfigureFlopFlip, ReconfigureFlopFlip, ToggleFeature, branchOnFeatureToggle, injectFeatureToggle, injectFeatureToggles, useAdapterReconfiguration, useAdapterStatus, useFeatureToggle, useFeatureToggles, useFlagVariation, useFlagVariations } from '@flopflip/react-broadcast';
export { default as ldAdapter } from '@flopflip/launchdarkly-adapter';
import 'prop-types';
import 'react';
import 'react-router-dom';
import '@babel/runtime-corejs3/helpers/defineProperty';
import '@emotion/styled/base';
import '@babel/runtime-corejs3/core-js-stable/instance/find';
import '@babel/runtime-corejs3/core-js-stable/instance/includes';
import '@babel/runtime-corejs3/core-js-stable/array/is-array';
import '@babel/runtime-corejs3/core-js-stable/object/keys';
import '@babel/runtime-corejs3/core-js-stable/object/get-own-property-symbols';
import '@babel/runtime-corejs3/core-js-stable/instance/filter';
import '@babel/runtime-corejs3/core-js-stable/object/get-own-property-descriptor';
import '@babel/runtime-corejs3/core-js-stable/instance/for-each';
import '@babel/runtime-corejs3/core-js-stable/object/get-own-property-descriptors';
import '@babel/runtime-corejs3/core-js-stable/object/define-properties';
import '@babel/runtime-corejs3/core-js-stable/object/define-property';
import '@emotion/react';
import '@commercetools-frontend/application-components';
import '@commercetools-frontend/constants';
import '@commercetools-frontend/i18n';
import '@commercetools-frontend/react-notifications';
import '@commercetools-frontend/sentry';
import './oidc-da8526e3.esm.js';
import '@babel/runtime-corejs3/core-js-stable/instance/concat';
import '@babel/runtime-corejs3/core-js-stable/instance/map';
import '@babel/runtime-corejs3/core-js-stable/url';
import '@commercetools-uikit/design-system';
import '@commercetools-uikit/spacings';
import 'react-intl';
import '@commercetools-uikit/flat-button';
import '@commercetools-uikit/icons';
import './use-is-served-by-proxy-77984220.esm.js';
import '@emotion/react/jsx-runtime';
import '@babel/runtime-corejs3/helpers/taggedTemplateLiteral';
import '@babel/runtime-corejs3/helpers/objectDestructuringEmpty';
import '@babel/runtime-corejs3/helpers/extends';
import 'memoize-one';
import 'react-select';
import '@commercetools-uikit/accessible-hidden';
import '@commercetools-uikit/select-input';
import '@commercetools-frontend/assets/images/ct-logo.svg';
import '@commercetools-uikit/loading-spinner';
import '@commercetools-frontend/browser-history';
import '@commercetools-frontend/l10n';
import '@babel/runtime-corejs3/helpers/slicedToArray';
import '@babel/runtime-corejs3/core-js-stable/reflect/construct';
import '@babel/runtime-corejs3/helpers/classCallCheck';
import '@babel/runtime-corejs3/helpers/createClass';
import '@babel/runtime-corejs3/helpers/inherits';
import '@babel/runtime-corejs3/helpers/possibleConstructorReturn';
import '@babel/runtime-corejs3/helpers/getPrototypeOf';
import '@babel/runtime-corejs3/core-js-stable/instance/index-of';
import '@babel/runtime-corejs3/core-js-stable/instance/slice';
import '@babel/runtime-corejs3/core-js-stable/object/from-entries';
import '@babel/runtime-corejs3/core-js-stable/instance/flags';
import '@apollo/client/react';
import '@flopflip/combine-adapters';
import '@flopflip/http-adapter';
import '@babel/runtime-corejs3/core-js-stable/instance/reduce';
import 'react-redux';
import 'lodash/upperFirst';
import '@babel/runtime-corejs3/core-js-stable/instance/starts-with';
import '@commercetools-uikit/design-system/materials/resets.css';
import '@commercetools-frontend/permissions';
import '@babel/runtime-corejs3/core-js-stable/json/stringify';
import '@babel/runtime-corejs3/core-js-stable/instance/some';
import '@commercetools-frontend/actions-global';
import '@babel/runtime-corejs3/core-js-stable/object/entries';
import '@babel/runtime-corejs3/core-js-stable/reflect/has';
import '@reduxjs/toolkit';
import 'lodash/mapValues';
import 'omit-empty-es';
import 'redux-thunk';
import '@commercetools-frontend/notifications';
import '@commercetools-frontend/sdk';
import 'redux-logger';
