// @commercetools-frontend/application-shell/dist/redirect-to-project-create-2266c427.cjs.prod.js
'use strict';

var react = require('@emotion/react');
var Card = require('@commercetools-uikit/card');
var designSystem = require('@commercetools-uikit/design-system');
var notifications = require('@commercetools-uikit/notifications');
var Spacings = require('@commercetools-uikit/spacings');
var Text = require('@commercetools-uikit/text');
var useIsServedByProxy = require('./use-is-served-by-proxy-845cace5.cjs.prod.js');
var jsxRuntime = require('@emotion/react/jsx-runtime');
require('@commercetools-frontend/application-shell-connectors');

function _interopDefault (e) { return e && e.__esModule ? e : { 'default': e }; }

var Card__default = /*#__PURE__*/_interopDefault(Card);
var Spacings__default = /*#__PURE__*/_interopDefault(Spacings);
var Text__default = /*#__PURE__*/_interopDefault(Text);

const RedirectToProjectCreate = () => {
  const servedByProxy = useIsServedByProxy.useIsServedByProxy();
  /**
   * NOTE:
   *   This looks a bit unusual: redirecting in render.
   *   However, when doing it in `cDM` we loose time
   *   we we actually do never want to render anything or
   *   interact with anything rendered. Instead we should should
   *   redirect away. Using a constructor would result in the same.
   *   In turn this intends wo make explicit that we never want to
   *   render and instead just navigate away.
   */
  if (servedByProxy === true) {
    useIsServedByProxy.location.replace('/account/projects/new');
    return null;
  }
  return jsxRuntime.jsx("div", {
    css: /*#__PURE__*/react.css("align-self:center;margin-top:", designSystem.customProperties.spacingXl, ";max-width:", designSystem.customProperties.constraint10, ";" + ("" ), "" ),
    children: jsxRuntime.jsx(Card__default["default"], {
      type: "flat",
      theme: "dark",
      children: jsxRuntime.jsxs(Spacings__default["default"].Stack, {
        children: [jsxRuntime.jsx(Text__default["default"].Headline, {
          as: "h1",
          children: "Please create a project!"
        }), jsxRuntime.jsx(notifications.ContentNotification, {
          type: "info",
          children: "You are running in development mode"
        }), jsxRuntime.jsx(Text__default["default"].Body, {
          children: "The Custom Application is not running behind the Merchant Center Proxy. Therefore, you are not being redirected to the account section to create a new project."
        }), jsxRuntime.jsx(Text__default["default"].Body, {
          children: "If you do need to create a project, we recommend to go to the Merchant Center production URL and create a project there. After that, you can access your new project from your local environment."
        })]
      })
    })
  });
};
RedirectToProjectCreate.displayName = 'RedirectToProjectCreate';

exports.RedirectToProjectCreate = RedirectToProjectCreate;
exports["default"] = RedirectToProjectCreate;
