// @commercetools-frontend/application-shell/dist/redirector-5daa8baa.cjs.dev.js
'use strict';

var _Object$keys = require('@babel/runtime-corejs3/core-js-stable/object/keys');
var _Object$getOwnPropertySymbols = require('@babel/runtime-corejs3/core-js-stable/object/get-own-property-symbols');
var _filterInstanceProperty = require('@babel/runtime-corejs3/core-js-stable/instance/filter');
var _Object$getOwnPropertyDescriptor = require('@babel/runtime-corejs3/core-js-stable/object/get-own-property-descriptor');
var _forEachInstanceProperty = require('@babel/runtime-corejs3/core-js-stable/instance/for-each');
var _Object$getOwnPropertyDescriptors = require('@babel/runtime-corejs3/core-js-stable/object/get-own-property-descriptors');
var _Object$defineProperties = require('@babel/runtime-corejs3/core-js-stable/object/define-properties');
var _Object$defineProperty = require('@babel/runtime-corejs3/core-js-stable/object/define-property');
var _defineProperty = require('@babel/runtime-corejs3/helpers/defineProperty');
var _pt = require('prop-types');
var _URL = require('@babel/runtime-corejs3/core-js-stable/url');
var react = require('react');
var qss = require('qss');
var useIsServedByProxy = require('./use-is-served-by-proxy-ed411f45.cjs.dev.js');

function _interopDefault (e) { return e && e.__esModule ? e : { 'default': e }; }

var _Object$keys__default = /*#__PURE__*/_interopDefault(_Object$keys);
var _Object$getOwnPropertySymbols__default = /*#__PURE__*/_interopDefault(_Object$getOwnPropertySymbols);
var _filterInstanceProperty__default = /*#__PURE__*/_interopDefault(_filterInstanceProperty);
var _Object$getOwnPropertyDescriptor__default = /*#__PURE__*/_interopDefault(_Object$getOwnPropertyDescriptor);
var _forEachInstanceProperty__default = /*#__PURE__*/_interopDefault(_forEachInstanceProperty);
var _Object$getOwnPropertyDescriptors__default = /*#__PURE__*/_interopDefault(_Object$getOwnPropertyDescriptors);
var _Object$defineProperties__default = /*#__PURE__*/_interopDefault(_Object$defineProperties);
var _Object$defineProperty__default = /*#__PURE__*/_interopDefault(_Object$defineProperty);
var _pt__default = /*#__PURE__*/_interopDefault(_pt);
var _URL__default = /*#__PURE__*/_interopDefault(_URL);

function ownKeys(e, r) { var t = _Object$keys__default["default"](e); if (_Object$getOwnPropertySymbols__default["default"]) { var o = _Object$getOwnPropertySymbols__default["default"](e); r && (o = _filterInstanceProperty__default["default"](o).call(o, function (r) { return _Object$getOwnPropertyDescriptor__default["default"](e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var _context, _context2; var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? _forEachInstanceProperty__default["default"](_context = ownKeys(Object(t), !0)).call(_context, function (r) { _defineProperty(e, r, t[r]); }) : _Object$getOwnPropertyDescriptors__default["default"] ? _Object$defineProperties__default["default"](e, _Object$getOwnPropertyDescriptors__default["default"](t)) : _forEachInstanceProperty__default["default"](_context2 = ownKeys(Object(t))).call(_context2, function (r) { _Object$defineProperty__default["default"](e, r, _Object$getOwnPropertyDescriptor__default["default"](t, r)); }); } return e; }
const redirectTo = targetUrl => {
  useIsServedByProxy.location.replace(targetUrl);
};
const Redirector = props => {
  var _props$origin, _props$queryParams, _enhancedLocation$que;
  // For now the authentication service runs on the same domain as the application,
  // even on development (using the webpack dev server).
  const originUrl = (_props$origin = props.origin) !== null && _props$origin !== void 0 ? _props$origin : window.location.origin;
  const enhancedLocation = props.location || {};
  const searchQuery = _objectSpread(_objectSpread({}, (_props$queryParams = props.queryParams) !== null && _props$queryParams !== void 0 ? _props$queryParams : {}), (_enhancedLocation$que = enhancedLocation.query) !== null && _enhancedLocation$que !== void 0 ? _enhancedLocation$que : {});
  const targetUrlObject = new _URL__default["default"](originUrl);
  targetUrlObject.pathname = props.to || targetUrlObject.pathname;
  targetUrlObject.search = "?".concat(qss.encode(searchQuery));
  const targetUrl = targetUrlObject.toString();
  react.useEffect(() => {
    redirectTo(targetUrl);
  }, [targetUrl]);
  return null;
};
Redirector.propTypes = {
  to: _pt__default["default"].string.isRequired,
  origin: _pt__default["default"].string,
  queryParams: _pt__default["default"].shape({
    redirectTo: _pt__default["default"].string,
    client_id: _pt__default["default"].string,
    response_type: _pt__default["default"].string,
    scope: _pt__default["default"].string,
    state: _pt__default["default"].string,
    nonce: _pt__default["default"].string
  })
};
Redirector.displayName = 'Redirector';

exports.Redirector = Redirector;
