// @commercetools-frontend/application-shell/dist/redirect-to-login-dffecd53.cjs.dev.js
'use strict';

var reactRouterDom = require('react-router-dom');
var uuid = require('uuid');
var applicationShellConnectors = require('@commercetools-frontend/application-shell-connectors');
var constants = require('@commercetools-frontend/constants');
var urlUtils = require('@commercetools-frontend/url-utils');
var oidc = require('./oidc-7c48215c.cjs.dev.js');
var useIsServedByProxy = require('./use-is-served-by-proxy-ed411f45.cjs.dev.js');
var getMcOrigin = require('./get-mc-origin-a86dff20.cjs.dev.js');
var redirector = require('./redirector-5daa8baa.cjs.dev.js');
var jsxRuntime = require('@emotion/react/jsx-runtime');
require('@babel/runtime-corejs3/core-js-stable/instance/concat');
require('@babel/runtime-corejs3/core-js-stable/instance/map');
require('@babel/runtime-corejs3/core-js-stable/instance/for-each');
require('@babel/runtime-corejs3/helpers/slicedToArray');
require('@babel/runtime-corejs3/core-js-stable/url');
require('@babel/runtime-corejs3/core-js-stable/instance/reverse');
require('@babel/runtime-corejs3/core-js-stable/object/keys');
require('@babel/runtime-corejs3/core-js-stable/object/get-own-property-symbols');
require('@babel/runtime-corejs3/core-js-stable/instance/filter');
require('@babel/runtime-corejs3/core-js-stable/object/get-own-property-descriptor');
require('@babel/runtime-corejs3/core-js-stable/object/get-own-property-descriptors');
require('@babel/runtime-corejs3/core-js-stable/object/define-properties');
require('@babel/runtime-corejs3/core-js-stable/object/define-property');
require('@babel/runtime-corejs3/helpers/defineProperty');
require('prop-types');
require('react');
require('qss');

const generateAndCacheNonceWithState = state => {
  const nonce = uuid.v4();
  // We store additional information within the given `nonce`
  // to then retrieve it later when the IdP redirects back
  // to our application. The URL will contain the `nonce` within
  // the id_token and, once validated, we can retrieve and use
  // the state object.
  // https://auth0.com/docs/protocols/oauth2/oauth-state#how-to-use-the-parameter-to-restore-application-state
  applicationShellConnectors.oidcStorage.setSessionState(nonce, state);
  return nonce;
};
const RedirectToLogin = () => {
  var _window$app$__DEVELOP;
  const location = reactRouterDom.useLocation();
  const servedByProxy = useIsServedByProxy.useIsServedByProxy();
  if ((_window$app$__DEVELOP = window.app.__DEVELOPMENT__) !== null && _window$app$__DEVELOP !== void 0 && (_window$app$__DEVELOP = _window$app$__DEVELOP.oidc) !== null && _window$app$__DEVELOP !== void 0 && _window$app$__DEVELOP.authorizeUrl) {
    var _window$app$__DEVELOP2, _window$app$__DEVELOP3, _window$app$__DEVELOP4, _window$app$__DEVELOP5, _window$app$__DEVELOP6;
    // We pick the project key from local storage. This assumes that the value
    // as been previously set when the application starts up.
    // This is necessary to allow switching projects and triggering a new login.
    const nextProjectKey = applicationShellConnectors.oidcStorage.getActiveProjectKey();

    // According to the OIDC spec, the `state` parameter is recommended to be sent
    // to the authorization server to help mitigating Cross-Site Request Forgery.
    // https://openid.net/specs/openid-connect-core-1_0.html#AuthRequest
    // Using `state` or `nonce` is very similar but with subtle differences.
    // Here we follow the approach described by Auth0 on how to use both, where
    // we generate a `nonce`, store it in session storage together with some state,
    // and send it to the authorization server as the `state` parameter.
    // https://auth0.com/docs/protocols/oauth2/redirect-users
    // Additionally, we still send the `nonce` parameter as well.
    const sessionId = generateAndCacheNonceWithState({
      applicationId: window.app.applicationId,
      // Store query parameters to be used after the callback redirect
      query: {
        redirectTo: nextProjectKey ? "/".concat(nextProjectKey) : '/'
      }
    });
    const requestedScope = oidc.buildOidcScope({
      projectKey: nextProjectKey !== null && nextProjectKey !== void 0 ? nextProjectKey : undefined,
      oAuthScopes: (_window$app$__DEVELOP2 = window.app.__DEVELOPMENT__) === null || _window$app$__DEVELOP2 === void 0 || (_window$app$__DEVELOP2 = _window$app$__DEVELOP2.oidc) === null || _window$app$__DEVELOP2 === void 0 ? void 0 : _window$app$__DEVELOP2.oAuthScopes,
      additionalOAuthScopes: (_window$app$__DEVELOP3 = window.app.__DEVELOPMENT__) === null || _window$app$__DEVELOP3 === void 0 || (_window$app$__DEVELOP3 = _window$app$__DEVELOP3.oidc) === null || _window$app$__DEVELOP3 === void 0 ? void 0 : _window$app$__DEVELOP3.additionalOAuthScopes,
      teamId: (_window$app$__DEVELOP4 = window.app.__DEVELOPMENT__) === null || _window$app$__DEVELOP4 === void 0 || (_window$app$__DEVELOP4 = _window$app$__DEVELOP4.oidc) === null || _window$app$__DEVELOP4 === void 0 ? void 0 : _window$app$__DEVELOP4.teamId,
      applicationId: (_window$app$__DEVELOP5 = window.app.__DEVELOPMENT__) === null || _window$app$__DEVELOP5 === void 0 || (_window$app$__DEVELOP5 = _window$app$__DEVELOP5.oidc) === null || _window$app$__DEVELOP5 === void 0 ? void 0 : _window$app$__DEVELOP5.applicationId
    });

    // Store session scopes, to be able to detect if requested scopes changed
    // in the application config and invalidate the session.
    // This is only valid for local development.
    applicationShellConnectors.oidcStorage.setSessionScope(requestedScope);
    return jsxRuntime.jsx(redirector.Redirector, {
      to: "",
      origin: (_window$app$__DEVELOP6 = window.app.__DEVELOPMENT__) === null || _window$app$__DEVELOP6 === void 0 || (_window$app$__DEVELOP6 = _window$app$__DEVELOP6.oidc) === null || _window$app$__DEVELOP6 === void 0 ? void 0 : _window$app$__DEVELOP6.authorizeUrl,
      location: location,
      queryParams: {
        reason: constants.LOGOUT_REASONS.UNAUTHORIZED,
        // Query parameters for OIDC-lik workflow.
        client_id: window.app.applicationId,
        response_type: oidc.OIDC_RESPONSE_TYPES.ID_TOKEN,
        scope: requestedScope,
        state: sessionId,
        nonce: sessionId
      }
    });
  }
  const mcOrigin = servedByProxy ? getMcOrigin.getMcOrigin(window.app.mcApiUrl) : undefined;
  return jsxRuntime.jsx(redirector.Redirector, {
    to: "login",
    origin: mcOrigin,
    location: location,
    queryParams: {
      reason: constants.LOGOUT_REASONS.UNAUTHORIZED,
      redirectTo: urlUtils.trimLeadingAndTrailingSlashes(urlUtils.joinPaths(window.location.origin, location.pathname))
    }
  });
};
RedirectToLogin.displayName = 'RedirectToLogin';

exports["default"] = RedirectToLogin;
