// @commercetools-frontend/application-shell/dist/service-page-project-switcher-698c4abe.cjs.prod.js
'use strict';

var _styled = require('@emotion/styled/base');
var reactRouterDom = require('react-router-dom');
var applicationShellConnectors = require('@commercetools-frontend/application-shell-connectors');
var dist_commercetoolsFrontendApplicationShell = require('./index-ae549827.cjs.prod.js');
var jsxRuntime = require('@emotion/react/jsx-runtime');

function _interopDefault (e) { return e && e.__esModule ? e : { 'default': e }; }

var _styled__default = /*#__PURE__*/_interopDefault(_styled);

const Container = /*#__PURE__*/_styled__default["default"]("div", {
  target: "efv14x0"
} )({
  name: "no8ssc",
  styles: "width:200px;text-align:left;margin:0 auto"
} );
const ServicePageProjectSwitcher = () => {
  const params = reactRouterDom.useParams();
  const numberOfProjects = applicationShellConnectors.useApplicationContext(context => context.user && context.user.projects.total || 0);
  if (numberOfProjects === 0) return null;
  return jsxRuntime.jsx(Container, {
    children: jsxRuntime.jsx(dist_commercetoolsFrontendApplicationShell.ProjectSwitcher
    // In this case it's not necessary to check if the `projectKey` param
    // is included in the list of projects. In such case
    // the dropdown will still be rendered but no project will be selected.
    // This is fine becase the user has still the possibility to "switch"
    // to a project.
    , {
      projectKey: params.projectKey
    })
  });
};
ServicePageProjectSwitcher.displayName = 'ServicePageProjectSwitcher';

exports.ServicePageProjectSwitcher = ServicePageProjectSwitcher;
