// @commercetools-frontend/application-shell/dist/application-entry-point-fd6f9973.cjs.dev.js
'use strict';

var _Object$keys = require('@babel/runtime-corejs3/core-js-stable/object/keys');
var _Object$getOwnPropertySymbols = require('@babel/runtime-corejs3/core-js-stable/object/get-own-property-symbols');
var _filterInstanceProperty = require('@babel/runtime-corejs3/core-js-stable/instance/filter');
var _Object$getOwnPropertyDescriptor = require('@babel/runtime-corejs3/core-js-stable/object/get-own-property-descriptor');
var _forEachInstanceProperty = require('@babel/runtime-corejs3/core-js-stable/instance/for-each');
var _Object$getOwnPropertyDescriptors = require('@babel/runtime-corejs3/core-js-stable/object/get-own-property-descriptors');
var _Object$defineProperties = require('@babel/runtime-corejs3/core-js-stable/object/define-properties');
var _Object$defineProperty = require('@babel/runtime-corejs3/core-js-stable/object/define-property');
var _defineProperty = require('@babel/runtime-corejs3/helpers/defineProperty');
var _pt = require('prop-types');
var react = require('react');
var reactRouterDom = require('react-router-dom');
var invariant = require('tiny-invariant');
var applicationComponents = require('@commercetools-frontend/application-components');
var ssr = require('@commercetools-frontend/application-config/ssr');
var permissions = require('@commercetools-frontend/permissions');
var index = require('./index-37a25e51.cjs.dev.js');
var jsxRuntime = require('@emotion/react/jsx-runtime');

function _interopDefault (e) { return e && e.__esModule ? e : { 'default': e }; }

var _Object$keys__default = /*#__PURE__*/_interopDefault(_Object$keys);
var _Object$getOwnPropertySymbols__default = /*#__PURE__*/_interopDefault(_Object$getOwnPropertySymbols);
var _filterInstanceProperty__default = /*#__PURE__*/_interopDefault(_filterInstanceProperty);
var _Object$getOwnPropertyDescriptor__default = /*#__PURE__*/_interopDefault(_Object$getOwnPropertyDescriptor);
var _forEachInstanceProperty__default = /*#__PURE__*/_interopDefault(_forEachInstanceProperty);
var _Object$getOwnPropertyDescriptors__default = /*#__PURE__*/_interopDefault(_Object$getOwnPropertyDescriptors);
var _Object$defineProperties__default = /*#__PURE__*/_interopDefault(_Object$defineProperties);
var _Object$defineProperty__default = /*#__PURE__*/_interopDefault(_Object$defineProperty);
var _pt__default = /*#__PURE__*/_interopDefault(_pt);
var invariant__default = /*#__PURE__*/_interopDefault(invariant);

function ownKeys(e, r) { var t = _Object$keys__default["default"](e); if (_Object$getOwnPropertySymbols__default["default"]) { var o = _Object$getOwnPropertySymbols__default["default"](e); r && (o = _filterInstanceProperty__default["default"](o).call(o, function (r) { return _Object$getOwnPropertyDescriptor__default["default"](e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var _context, _context2; var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? _forEachInstanceProperty__default["default"](_context = ownKeys(Object(t), !0)).call(_context, function (r) { _defineProperty(e, r, t[r]); }) : _Object$getOwnPropertyDescriptors__default["default"] ? _Object$defineProperties__default["default"](e, _Object$getOwnPropertyDescriptors__default["default"](t)) : _forEachInstanceProperty__default["default"](_context2 = ownKeys(Object(t))).call(_context2, function (r) { _Object$defineProperty__default["default"](e, r, _Object$getOwnPropertyDescriptor__default["default"](t, r)); }); } return e; }
const ApplicationRouteWithPermissionCheck = props => {
  const permissionKeys = ssr.entryPointUriPathToPermissionKeys(props.environment.entryPointUriPath);
  // Require View permission to render the application.
  const canView = permissions.useIsAuthorized({
    demandedPermissions: [permissionKeys.View]
  });
  if (canView) {
    return jsxRuntime.jsx(jsxRuntime.Fragment, {
      children: react.Children.only(props.children)
    });
  }
  return jsxRuntime.jsx(applicationComponents.PageUnauthorized, {});
};
ApplicationRouteWithPermissionCheck.propTypes = process.env.NODE_ENV !== "production" ? {
  disableRoutePermissionCheck: _pt__default["default"].bool,
  render: _pt__default["default"].func,
  children: _pt__default["default"].node
} : {};
const ApplicationRoute = props => {
  if (props.disableRoutePermissionCheck) {
    return jsxRuntime.jsx(jsxRuntime.Fragment, {
      children: react.Children.only(props.children)
    });
  }
  return jsxRuntime.jsx(ApplicationRouteWithPermissionCheck, _objectSpread({}, props));
};
ApplicationRoute.propTypes = process.env.NODE_ENV !== "production" ? {
  disableRoutePermissionCheck: _pt__default["default"].bool,
  render: _pt__default["default"].func,
  children: _pt__default["default"].node
} : {};
const ApplicationEntryPoint = props => {
  // If the `children` prop is used (instead of the `render` prop),
  // we pre-configure the application entry point routes to avoid
  // users to do so on their own.
  if (props.children) {
    const entryPointUriPath = props.environment.entryPointUriPath;
    return jsxRuntime.jsxs(reactRouterDom.Switch, {
      children: [
      // For development, it's useful to redirect to the actual
      // application routes when you open the browser at http://localhost:3001
      process.env.NODE_ENV === 'production' ? null : jsxRuntime.jsx(reactRouterDom.Redirect, {
        exact: true,
        from: "/:projectKey",
        to: "/:projectKey/".concat(entryPointUriPath)
      }), jsxRuntime.jsx(reactRouterDom.Route, {
        path: "/:projectKey/".concat(entryPointUriPath),
        children: jsxRuntime.jsx(ApplicationRoute, _objectSpread({}, props))
      }), jsxRuntime.jsx(index.RouteCatchAll, {})]
    });
  }

  // We still support the `render` prop, for backwards compatibility
  // and for having more control in certain cases (mostly to some of our internal apps).
  if (props.render && typeof props.render === 'function') {
    return jsxRuntime.jsx(jsxRuntime.Fragment, {
      children: props.render()
    });
  }

  // The `render` prop function is still required (backwards compatibility).
  !!props.render ? process.env.NODE_ENV !== "production" ? invariant__default["default"](false, '@commercetools-frontend/application-shell: Missing required function prop "render".') : invariant__default["default"](false) : void 0;
  return null;
};
ApplicationEntryPoint.propTypes = process.env.NODE_ENV !== "production" ? {
  disableRoutePermissionCheck: _pt__default["default"].bool,
  render: _pt__default["default"].func,
  children: _pt__default["default"].node
} : {};

exports.ApplicationEntryPoint = ApplicationEntryPoint;
