// @commercetools-frontend/application-shell/dist/commercetools-frontend-application-shell.cjs.dev.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var dist_commercetoolsFrontendApplicationShell = require('./index-9c9f0152.cjs.dev.js');
var index = require('./index-37a25e51.cjs.dev.js');
var applicationShellConnectors = require('@commercetools-frontend/application-shell-connectors');
var ssr = require('@commercetools-frontend/application-config/ssr');
var reactBroadcast = require('@flopflip/react-broadcast');
var ldAdapter = require('@flopflip/launchdarkly-adapter');
require('prop-types');
require('react');
require('react-router-dom');
require('@babel/runtime-corejs3/helpers/defineProperty');
require('@emotion/styled/base');
require('@babel/runtime-corejs3/core-js-stable/instance/find');
require('@babel/runtime-corejs3/core-js-stable/instance/includes');
require('@babel/runtime-corejs3/core-js-stable/array/is-array');
require('@babel/runtime-corejs3/core-js-stable/object/keys');
require('@babel/runtime-corejs3/core-js-stable/object/get-own-property-symbols');
require('@babel/runtime-corejs3/core-js-stable/instance/filter');
require('@babel/runtime-corejs3/core-js-stable/object/get-own-property-descriptor');
require('@babel/runtime-corejs3/core-js-stable/instance/for-each');
require('@babel/runtime-corejs3/core-js-stable/object/get-own-property-descriptors');
require('@babel/runtime-corejs3/core-js-stable/object/define-properties');
require('@babel/runtime-corejs3/core-js-stable/object/define-property');
require('@emotion/react');
require('@commercetools-frontend/application-components');
require('@commercetools-frontend/constants');
require('@commercetools-frontend/i18n');
require('@commercetools-frontend/react-notifications');
require('@commercetools-frontend/sentry');
require('./oidc-7c48215c.cjs.dev.js');
require('@babel/runtime-corejs3/core-js-stable/instance/concat');
require('@babel/runtime-corejs3/core-js-stable/instance/map');
require('@babel/runtime-corejs3/core-js-stable/url');
require('@commercetools-uikit/design-system');
require('@commercetools-uikit/spacings');
require('react-intl');
require('@commercetools-uikit/flat-button');
require('@commercetools-uikit/icons');
require('./use-is-served-by-proxy-ed411f45.cjs.dev.js');
require('@emotion/react/jsx-runtime');
require('@babel/runtime-corejs3/helpers/taggedTemplateLiteral');
require('@babel/runtime-corejs3/helpers/objectDestructuringEmpty');
require('@babel/runtime-corejs3/helpers/extends');
require('memoize-one');
require('react-select');
require('@commercetools-uikit/accessible-hidden');
require('@commercetools-uikit/select-input');
require('@commercetools-frontend/assets/images/ct-logo.svg');
require('@commercetools-uikit/loading-spinner');
require('@commercetools-frontend/browser-history');
require('@commercetools-frontend/l10n');
require('@babel/runtime-corejs3/helpers/slicedToArray');
require('@babel/runtime-corejs3/core-js-stable/reflect/construct');
require('@babel/runtime-corejs3/helpers/classCallCheck');
require('@babel/runtime-corejs3/helpers/createClass');
require('@babel/runtime-corejs3/helpers/inherits');
require('@babel/runtime-corejs3/helpers/possibleConstructorReturn');
require('@babel/runtime-corejs3/helpers/getPrototypeOf');
require('@babel/runtime-corejs3/core-js-stable/instance/index-of');
require('@babel/runtime-corejs3/core-js-stable/instance/slice');
require('@babel/runtime-corejs3/core-js-stable/object/from-entries');
require('@babel/runtime-corejs3/core-js-stable/instance/flags');
require('@apollo/client/react');
require('@flopflip/combine-adapters');
require('@flopflip/http-adapter');
require('@babel/runtime-corejs3/core-js-stable/instance/reduce');
require('react-redux');
require('lodash/upperFirst');
require('@babel/runtime-corejs3/core-js-stable/instance/starts-with');
require('@commercetools-uikit/design-system/materials/resets.css');
require('@commercetools-frontend/permissions');
require('@babel/runtime-corejs3/core-js-stable/json/stringify');
require('@babel/runtime-corejs3/core-js-stable/instance/some');
require('@commercetools-frontend/actions-global');
require('@babel/runtime-corejs3/core-js-stable/object/entries');
require('@babel/runtime-corejs3/core-js-stable/reflect/has');
require('@reduxjs/toolkit');
require('lodash/mapValues');
require('omit-empty-es');
require('redux-thunk');
require('@commercetools-frontend/notifications');
require('@commercetools-frontend/sdk');
require('redux-logger');

function _interopDefault (e) { return e && e.__esModule ? e : { 'default': e }; }

var ldAdapter__default = /*#__PURE__*/_interopDefault(ldAdapter);



exports.ApplicationPageTitle = dist_commercetoolsFrontendApplicationShell.ApplicationPageTitle;
exports.ApplicationShell = dist_commercetoolsFrontendApplicationShell.ApplicationShell;
exports.ApplicationShellProvider = dist_commercetoolsFrontendApplicationShell.ApplicationShellProvider;
exports.ConfigureIntlProvider = dist_commercetoolsFrontendApplicationShell.ConfigureIntlProvider;
exports.CustomViewShell = dist_commercetoolsFrontendApplicationShell.CustomViewShellWrapper;
exports.InjectReducers = dist_commercetoolsFrontendApplicationShell.InjectReducers;
exports.SetupFlopFlipProvider = dist_commercetoolsFrontendApplicationShell.SetupFlopFlipProvider;
exports.SuspendedRoute = dist_commercetoolsFrontendApplicationShell.SuspendedRoute;
exports.setupGlobalErrorListener = dist_commercetoolsFrontendApplicationShell.setupGlobalErrorListener;
exports.useRoutesCreator = dist_commercetoolsFrontendApplicationShell.useRoutesCreator;
exports.version = dist_commercetoolsFrontendApplicationShell.applicationShellVersion;
exports.RouteCatchAll = index.RouteCatchAll;
exports.applyDefaultMiddlewares = index.applyDefaultMiddlewares;
Object.defineProperty(exports, 'buildApiUrl', {
	enumerable: true,
	get: function () { return applicationShellConnectors.buildApiUrl; }
});
Object.defineProperty(exports, 'createApolloClient', {
	enumerable: true,
	get: function () { return applicationShellConnectors.createApolloClient; }
});
Object.defineProperty(exports, 'createApolloContextForProxyForwardTo', {
	enumerable: true,
	get: function () { return applicationShellConnectors.createApolloContextForProxyForwardTo; }
});
Object.defineProperty(exports, 'createHttpClientOptions', {
	enumerable: true,
	get: function () { return applicationShellConnectors.createHttpClientOptions; }
});
Object.defineProperty(exports, 'executeHttpClientRequest', {
	enumerable: true,
	get: function () { return applicationShellConnectors.executeHttpClientRequest; }
});
Object.defineProperty(exports, 'getMcApiUrl', {
	enumerable: true,
	get: function () { return applicationShellConnectors.getMcApiUrl; }
});
Object.defineProperty(exports, 'selectProjectKeyFromUrl', {
	enumerable: true,
	get: function () { return applicationShellConnectors.selectProjectKeyFromUrl; }
});
Object.defineProperty(exports, 'selectUserId', {
	enumerable: true,
	get: function () { return applicationShellConnectors.selectUserId; }
});
Object.defineProperty(exports, 'useMcLazyQuery', {
	enumerable: true,
	get: function () { return applicationShellConnectors.useMcLazyQuery; }
});
Object.defineProperty(exports, 'useMcMutation', {
	enumerable: true,
	get: function () { return applicationShellConnectors.useMcMutation; }
});
Object.defineProperty(exports, 'useMcQuery', {
	enumerable: true,
	get: function () { return applicationShellConnectors.useMcQuery; }
});
Object.defineProperty(exports, 'entryPointUriPathToPermissionKeys', {
	enumerable: true,
	get: function () { return ssr.entryPointUriPathToPermissionKeys; }
});
Object.defineProperty(exports, 'entryPointUriPathToResourceAccesses', {
	enumerable: true,
	get: function () { return ssr.entryPointUriPathToResourceAccesses; }
});
Object.defineProperty(exports, 'ConfigureFlopFlip', {
	enumerable: true,
	get: function () { return reactBroadcast.ConfigureFlopFlip; }
});
Object.defineProperty(exports, 'ReconfigureFlopFlip', {
	enumerable: true,
	get: function () { return reactBroadcast.ReconfigureFlopFlip; }
});
Object.defineProperty(exports, 'ToggleFeature', {
	enumerable: true,
	get: function () { return reactBroadcast.ToggleFeature; }
});
Object.defineProperty(exports, 'branchOnFeatureToggle', {
	enumerable: true,
	get: function () { return reactBroadcast.branchOnFeatureToggle; }
});
Object.defineProperty(exports, 'injectFeatureToggle', {
	enumerable: true,
	get: function () { return reactBroadcast.injectFeatureToggle; }
});
Object.defineProperty(exports, 'injectFeatureToggles', {
	enumerable: true,
	get: function () { return reactBroadcast.injectFeatureToggles; }
});
Object.defineProperty(exports, 'useAdapterReconfiguration', {
	enumerable: true,
	get: function () { return reactBroadcast.useAdapterReconfiguration; }
});
Object.defineProperty(exports, 'useAdapterStatus', {
	enumerable: true,
	get: function () { return reactBroadcast.useAdapterStatus; }
});
Object.defineProperty(exports, 'useFeatureToggle', {
	enumerable: true,
	get: function () { return reactBroadcast.useFeatureToggle; }
});
Object.defineProperty(exports, 'useFeatureToggles', {
	enumerable: true,
	get: function () { return reactBroadcast.useFeatureToggles; }
});
Object.defineProperty(exports, 'useFlagVariation', {
	enumerable: true,
	get: function () { return reactBroadcast.useFlagVariation; }
});
Object.defineProperty(exports, 'useFlagVariations', {
	enumerable: true,
	get: function () { return reactBroadcast.useFlagVariations; }
});
Object.defineProperty(exports, 'ldAdapter', {
	enumerable: true,
	get: function () { return ldAdapter__default["default"]; }
});
