// @commercetools-frontend/application-shell/dist/custom-view-dev-host-7060a2b1.cjs.dev.js
'use strict';

var _slicedToArray = require('@babel/runtime-corejs3/helpers/slicedToArray');
var _pt = require('prop-types');
var _mapInstanceProperty = require('@babel/runtime-corejs3/core-js-stable/instance/map');
var _concatInstanceProperty = require('@babel/runtime-corejs3/core-js-stable/instance/concat');
var react = require('react');
var reactRouterDom = require('react-router-dom');
var applicationComponents = require('@commercetools-frontend/application-components');
var history = require('@commercetools-frontend/browser-history');
var Constraints = require('@commercetools-uikit/constraints');
var PrimaryButton = require('@commercetools-uikit/primary-button');
var Spacings = require('@commercetools-uikit/spacings');
var Text = require('@commercetools-uikit/text');
var dist_commercetoolsFrontendApplicationShell = require('./index-9c9f0152.cjs.dev.js');
var jsxRuntime = require('@emotion/react/jsx-runtime');
require('./index-37a25e51.cjs.dev.js');
require('@babel/runtime-corejs3/core-js-stable/object/keys');
require('@babel/runtime-corejs3/core-js-stable/object/get-own-property-symbols');
require('@babel/runtime-corejs3/core-js-stable/instance/filter');
require('@babel/runtime-corejs3/core-js-stable/object/get-own-property-descriptor');
require('@babel/runtime-corejs3/core-js-stable/instance/for-each');
require('@babel/runtime-corejs3/core-js-stable/object/get-own-property-descriptors');
require('@babel/runtime-corejs3/core-js-stable/object/define-properties');
require('@babel/runtime-corejs3/core-js-stable/object/define-property');
require('@babel/runtime-corejs3/helpers/defineProperty');
require('@babel/runtime-corejs3/core-js-stable/instance/reduce');
require('@babel/runtime-corejs3/core-js-stable/object/entries');
require('@babel/runtime-corejs3/core-js-stable/reflect/has');
require('@reduxjs/toolkit');
require('lodash/mapValues');
require('omit-empty-es');
require('redux-thunk');
require('@commercetools-frontend/application-shell-connectors');
require('@commercetools-frontend/constants');
require('@commercetools-frontend/notifications');
require('@commercetools-frontend/sdk');
require('@babel/runtime-corejs3/core-js-stable/instance/index-of');
require('@babel/runtime-corejs3/core-js-stable/instance/slice');
require('@babel/runtime-corejs3/core-js-stable/instance/includes');
require('@commercetools-frontend/sentry');
require('@commercetools-frontend/react-notifications');
require('@babel/runtime-corejs3/core-js-stable/instance/starts-with');
require('redux-logger');
require('@emotion/styled/base');
require('@babel/runtime-corejs3/core-js-stable/instance/find');
require('@babel/runtime-corejs3/core-js-stable/array/is-array');
require('@emotion/react');
require('@commercetools-frontend/i18n');
require('./oidc-7c48215c.cjs.dev.js');
require('@babel/runtime-corejs3/core-js-stable/url');
require('@commercetools-uikit/design-system');
require('react-intl');
require('@commercetools-uikit/flat-button');
require('@commercetools-uikit/icons');
require('./use-is-served-by-proxy-ed411f45.cjs.dev.js');
require('@babel/runtime-corejs3/helpers/taggedTemplateLiteral');
require('@babel/runtime-corejs3/helpers/objectDestructuringEmpty');
require('@babel/runtime-corejs3/helpers/extends');
require('memoize-one');
require('react-select');
require('@commercetools-uikit/accessible-hidden');
require('@commercetools-uikit/select-input');
require('@commercetools-frontend/assets/images/ct-logo.svg');
require('@commercetools-uikit/loading-spinner');
require('@commercetools-frontend/l10n');
require('@babel/runtime-corejs3/core-js-stable/reflect/construct');
require('@babel/runtime-corejs3/helpers/classCallCheck');
require('@babel/runtime-corejs3/helpers/createClass');
require('@babel/runtime-corejs3/helpers/inherits');
require('@babel/runtime-corejs3/helpers/possibleConstructorReturn');
require('@babel/runtime-corejs3/helpers/getPrototypeOf');
require('@babel/runtime-corejs3/core-js-stable/object/from-entries');
require('@babel/runtime-corejs3/core-js-stable/instance/flags');
require('@apollo/client/react');
require('@flopflip/combine-adapters');
require('@flopflip/http-adapter');
require('@flopflip/launchdarkly-adapter');
require('@flopflip/react-broadcast');
require('react-redux');
require('lodash/upperFirst');
require('@commercetools-uikit/design-system/materials/resets.css');
require('@commercetools-frontend/application-config/ssr');
require('@commercetools-frontend/permissions');
require('@babel/runtime-corejs3/core-js-stable/json/stringify');
require('@babel/runtime-corejs3/core-js-stable/instance/some');
require('@commercetools-frontend/actions-global');

function _interopDefault (e) { return e && e.__esModule ? e : { 'default': e }; }

var _pt__default = /*#__PURE__*/_interopDefault(_pt);
var _mapInstanceProperty__default = /*#__PURE__*/_interopDefault(_mapInstanceProperty);
var _concatInstanceProperty__default = /*#__PURE__*/_interopDefault(_concatInstanceProperty);
var history__default = /*#__PURE__*/_interopDefault(history);
var Constraints__default = /*#__PURE__*/_interopDefault(Constraints);
var PrimaryButton__default = /*#__PURE__*/_interopDefault(PrimaryButton);
var Spacings__default = /*#__PURE__*/_interopDefault(Spacings);
var Text__default = /*#__PURE__*/_interopDefault(Text);

const LocalCustomViewLauncher = props => {
  var _props$environment$__, _context, _customViewConfig$typ, _props$environment$__2;
  const _useState = react.useState(false),
    _useState2 = _slicedToArray(_useState, 2),
    shouldRenderCustomView = _useState2[0],
    setShouldRenderCustomView = _useState2[1];
  // We expect this in development to be defined.
  const customViewConfig = (_props$environment$__ = props.environment.__DEVELOPMENT__) === null || _props$environment$__ === void 0 ? void 0 : _props$environment$__.customViewConfig;
  return jsxRuntime.jsx("main", {
    children: jsxRuntime.jsx(Constraints__default["default"].Horizontal, {
      max: "scale",
      children: jsxRuntime.jsxs(Spacings__default["default"].Inset, {
        scale: "xl",
        children: [jsxRuntime.jsxs(Spacings__default["default"].Stack, {
          scale: "xl",
          children: [jsxRuntime.jsx(Text__default["default"].Headline, {
            as: "h1",
            children: "Custom View loader"
          }), jsxRuntime.jsxs(Spacings__default["default"].Stack, {
            scale: "l",
            children: [jsxRuntime.jsxs(Spacings__default["default"].Stack, {
              scale: "m",
              children: [jsxRuntime.jsxs(Text__default["default"].Body, {
                children: ["This page simulates the rendering of your Custom View as if it was within a page in the Merchant Center:", jsxRuntime.jsx("br", {}), "Based on your ", jsxRuntime.jsx("i", {
                  children: "custom-view-config"
                }), " file, these are the settings that will be used:"]
              }), jsxRuntime.jsx("ul", {
                children: jsxRuntime.jsxs(Spacings__default["default"].Stack, {
                  scale: "s",
                  children: [jsxRuntime.jsxs("li", {
                    children: ["Custom View ID: ", jsxRuntime.jsx("b", {
                      children: customViewConfig.id
                    })]
                  }), jsxRuntime.jsxs("li", {
                    children: ["Custom View permissions:", ' ', jsxRuntime.jsx("b", {
                      children: _mapInstanceProperty__default["default"](_context = customViewConfig.permissions).call(_context, permission => {
                        var _context2;
                        return _concatInstanceProperty__default["default"](_context2 = "(".concat(permission.name, ": ")).call(_context2, permission.oAuthScopes.join(', '), ")");
                      }).join(', ')
                    })]
                  }), jsxRuntime.jsxs("li", {
                    children: ["Custom View locators:", ' ', jsxRuntime.jsx("b", {
                      children: customViewConfig.locators.join(', ')
                    })]
                  }), jsxRuntime.jsxs("li", {
                    children: ["Custom View type: ", jsxRuntime.jsx("b", {
                      children: customViewConfig.type
                    })]
                  }), jsxRuntime.jsxs("li", {
                    children: ["Custom View size:", ' ', jsxRuntime.jsx("b", {
                      children: (_customViewConfig$typ = customViewConfig.typeSettings) === null || _customViewConfig$typ === void 0 ? void 0 : _customViewConfig$typ.size
                    })]
                  })]
                })
              })]
            }), jsxRuntime.jsx(Constraints__default["default"].Horizontal, {
              max: 10,
              children: jsxRuntime.jsx(PrimaryButton__default["default"], {
                label: "Open the Custom View",
                onClick: () => setShouldRenderCustomView(true)
              })
            })]
          })]
        }), shouldRenderCustomView && jsxRuntime.jsx(applicationComponents.CustomViewLoader, {
          customView: customViewConfig,
          hostUrl: (_props$environment$__2 = props.environment.__DEVELOPMENT__) === null || _props$environment$__2 === void 0 ? void 0 : _props$environment$__2.customViewHostUrl,
          onClose: () => setShouldRenderCustomView(false)
        })]
      })
    })
  });
};
const SimulatedIframeRoute = props => {
  const routeMatch = reactRouterDom.useRouteMatch();
  console.info("\u2139\uFE0F Rendering a Custom View as it would be rendered within an iframe", routeMatch.url);
  return jsxRuntime.jsx(jsxRuntime.Fragment, {
    children: props.children
  });
};
const CustomViewDevHost = props => {
  return jsxRuntime.jsx(reactRouterDom.Router, {
    history: history__default["default"],
    children: jsxRuntime.jsxs(reactRouterDom.Switch, {
      children: [jsxRuntime.jsx(reactRouterDom.Route, {
        path: "/custom-views/:customViewId/projects/:projectKey",
        children: jsxRuntime.jsx(SimulatedIframeRoute, {
          children: props.children
        })
      }), jsxRuntime.jsx(reactRouterDom.Route, {
        children: jsxRuntime.jsx(dist_commercetoolsFrontendApplicationShell.ApplicationShell, {
          environment: window.app,
          applicationMessages: props.applicationMessages,
          disableRoutePermissionCheck: true,
          children: jsxRuntime.jsx(LocalCustomViewLauncher, {
            environment: window.app
          })
        })
      })]
    })
  });
};
CustomViewDevHost.propTypes = process.env.NODE_ENV !== "production" ? {
  children: _pt__default["default"].node.isRequired
} : {};
CustomViewDevHost.displayName = 'CustomViewDevHost';

exports["default"] = CustomViewDevHost;
