// @commercetools-frontend/application-shell/dist/index-6cd618cc.cjs.prod.js
'use strict';

var _Object$keys = require('@babel/runtime-corejs3/core-js-stable/object/keys');
var _Object$getOwnPropertySymbols = require('@babel/runtime-corejs3/core-js-stable/object/get-own-property-symbols');
var _filterInstanceProperty = require('@babel/runtime-corejs3/core-js-stable/instance/filter');
var _Object$getOwnPropertyDescriptor = require('@babel/runtime-corejs3/core-js-stable/object/get-own-property-descriptor');
var _forEachInstanceProperty = require('@babel/runtime-corejs3/core-js-stable/instance/for-each');
var _Object$getOwnPropertyDescriptors = require('@babel/runtime-corejs3/core-js-stable/object/get-own-property-descriptors');
var _Object$defineProperties = require('@babel/runtime-corejs3/core-js-stable/object/define-properties');
var _Object$defineProperty = require('@babel/runtime-corejs3/core-js-stable/object/define-property');
var _defineProperty = require('@babel/runtime-corejs3/helpers/defineProperty');
var _slicedToArray = require('@babel/runtime-corejs3/helpers/slicedToArray');
var _reduceInstanceProperty = require('@babel/runtime-corejs3/core-js-stable/instance/reduce');
var _Object$entries = require('@babel/runtime-corejs3/core-js-stable/object/entries');
var _Reflect$has = require('@babel/runtime-corejs3/core-js-stable/reflect/has');
var toolkit = require('@reduxjs/toolkit');
var mapValues = require('lodash/mapValues');
var omitEmpty = require('omit-empty-es');
var thunk = require('redux-thunk');
var applicationShellConnectors = require('@commercetools-frontend/application-shell-connectors');
var constants = require('@commercetools-frontend/constants');
var notifications = require('@commercetools-frontend/notifications');
var sdk = require('@commercetools-frontend/sdk');
var react = require('react');
var _indexOfInstanceProperty = require('@babel/runtime-corejs3/core-js-stable/instance/index-of');
var _sliceInstanceProperty = require('@babel/runtime-corejs3/core-js-stable/instance/slice');
var _includesInstanceProperty = require('@babel/runtime-corejs3/core-js-stable/instance/includes');
var sentry = require('@commercetools-frontend/sentry');
var reactNotifications = require('@commercetools-frontend/react-notifications');
var _startsWithInstanceProperty = require('@babel/runtime-corejs3/core-js-stable/instance/starts-with');
var reduxLogger = require('redux-logger');

function _interopDefault (e) { return e && e.__esModule ? e : { 'default': e }; }

var _Object$keys__default = /*#__PURE__*/_interopDefault(_Object$keys);
var _Object$getOwnPropertySymbols__default = /*#__PURE__*/_interopDefault(_Object$getOwnPropertySymbols);
var _filterInstanceProperty__default = /*#__PURE__*/_interopDefault(_filterInstanceProperty);
var _Object$getOwnPropertyDescriptor__default = /*#__PURE__*/_interopDefault(_Object$getOwnPropertyDescriptor);
var _forEachInstanceProperty__default = /*#__PURE__*/_interopDefault(_forEachInstanceProperty);
var _Object$getOwnPropertyDescriptors__default = /*#__PURE__*/_interopDefault(_Object$getOwnPropertyDescriptors);
var _Object$defineProperties__default = /*#__PURE__*/_interopDefault(_Object$defineProperties);
var _Object$defineProperty__default = /*#__PURE__*/_interopDefault(_Object$defineProperty);
var _reduceInstanceProperty__default = /*#__PURE__*/_interopDefault(_reduceInstanceProperty);
var _Object$entries__default = /*#__PURE__*/_interopDefault(_Object$entries);
var _Reflect$has__default = /*#__PURE__*/_interopDefault(_Reflect$has);
var mapValues__default = /*#__PURE__*/_interopDefault(mapValues);
var omitEmpty__default = /*#__PURE__*/_interopDefault(omitEmpty);
var thunk__default = /*#__PURE__*/_interopDefault(thunk);
var _indexOfInstanceProperty__default = /*#__PURE__*/_interopDefault(_indexOfInstanceProperty);
var _sliceInstanceProperty__default = /*#__PURE__*/_interopDefault(_sliceInstanceProperty);
var _includesInstanceProperty__default = /*#__PURE__*/_interopDefault(_includesInstanceProperty);
var _startsWithInstanceProperty__default = /*#__PURE__*/_interopDefault(_startsWithInstanceProperty);

const isShowRequestInFlightAction = action => action.type === constants.SHOW_LOADING;
const isHideRequestInFlightAction = action => action.type === constants.HIDE_LOADING;
const excludeFirstOccurrence = (list, item) => {
  const index = _indexOfInstanceProperty__default["default"](list).call(list, item);
  return [..._sliceInstanceProperty__default["default"](list).call(list, 0, index), ..._sliceInstanceProperty__default["default"](list).call(list, index + 1)];
};
var requestsInFlightReducer = ((requestsInFlight, action) => {
  if (!requestsInFlight || !action) return [];
  if (isShowRequestInFlightAction(action)) return [...requestsInFlight, action.payload];
  if (isHideRequestInFlightAction(action)) {
    // may only remove first occurence
    if (!_includesInstanceProperty__default["default"](requestsInFlight).call(requestsInFlight, action.payload)) {
      sentry.reportErrorToSentry(new Error("Tried to hide \"".concat(action.payload, "\", but it was not progressing!")));
      return requestsInFlight;
    }
    return excludeFirstOccurrence(requestsInFlight, action.payload);
  }
  return requestsInFlight;
});

const isAddNotificationErrorAction = action => {
  const errorAction = action;
  return errorAction.type === notifications.ADD_NOTIFICATION && errorAction.payload && (errorAction.payload.kind === constants.NOTIFICATION_KINDS_PAGE['api-error'] || errorAction.payload.kind === constants.NOTIFICATION_KINDS_PAGE['unexpected-error']);
};
const isHideAllNotificationsAction = action => action.type === constants.HIDE_ALL_PAGE_NOTIFICATIONS;
const hideNotificationsMiddleware = _ref => {
  let getState = _ref.getState;
  return next => action => {
    if (isHideAllNotificationsAction(action) || isAddNotificationErrorAction(action)) {
      var _context;
      const state = getState();
      _forEachInstanceProperty__default["default"](_context = reactNotifications.selectPageNotifications(state)).call(_context, notification => {
        next(notifications.removeNotification(notification.id));
      });
    }
    return next(action);
  };
};

const loggerMiddleware = reduxLogger.createLogger({
  logger: applicationShellConnectors.logger,
  collapsed: true,
  colors: {
    title: () => '#000000',
    prevState: () => '#9E9E9E',
    action: () => '#03A9F4',
    nextState: () => '#4CAF50',
    error: () => '#F20404'
  },
  // WARNING: Enabling this option causes huge performance degradation.
  // Only enable if you want detailed debugging.
  diff: false,
  predicate: (_getState, action) => {
    if (!action) return false;
    const type = action.type,
      payload = action.payload;
    if (!type) return false;
    if (_startsWithInstanceProperty__default["default"](type).call(type, 'SHOW_') || _startsWithInstanceProperty__default["default"](type).call(type, 'HIDE_') || _startsWithInstanceProperty__default["default"](type).call(type, 'APOLLO_')) return false;
    switch (type) {
      case notifications.REMOVE_NOTIFICATION:
        return false;
      case notifications.ADD_NOTIFICATION:
        {
          const ignoredKinds = ['api-error', 'unexpected-error'];
          return !_includesInstanceProperty__default["default"](ignoredKinds).call(ignoredKinds, payload.kind);
        }
      default:
        return true;
    }
  }
});

function ownKeys(e, r) { var t = _Object$keys__default["default"](e); if (_Object$getOwnPropertySymbols__default["default"]) { var o = _Object$getOwnPropertySymbols__default["default"](e); r && (o = _filterInstanceProperty__default["default"](o).call(o, function (r) { return _Object$getOwnPropertyDescriptor__default["default"](e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var _context2, _context3; var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? _forEachInstanceProperty__default["default"](_context2 = ownKeys(Object(t), !0)).call(_context2, function (r) { _defineProperty(e, r, t[r]); }) : _Object$getOwnPropertyDescriptors__default["default"] ? _Object$defineProperties__default["default"](e, _Object$getOwnPropertyDescriptors__default["default"](t)) : _forEachInstanceProperty__default["default"](_context3 = ownKeys(Object(t))).call(_context3, function (r) { _Object$defineProperty__default["default"](e, r, _Object$getOwnPropertyDescriptor__default["default"](t, r)); }); } return e; }
const mergeObjectValues = object => {
  var _context;
  return _reduceInstanceProperty__default["default"](_context = _Object$entries__default["default"](object)).call(_context, (acc, _ref) => {
    let _ref2 = _slicedToArray(_ref, 2),
      value = _ref2[1];
    return _objectSpread(_objectSpread({}, acc), value);
  }, {});
};
const patchedGetCorrelationId = () => applicationShellConnectors.getCorrelationId({
  userId: applicationShellConnectors.selectUserId()
});
const getAdditionalHeaders = () => {
  const sessionToken = applicationShellConnectors.oidcStorage.getSessionToken();
  return omitEmpty__default["default"]({
    [constants.SUPPORTED_HEADERS.AUTHORIZATION]: sessionToken ? "Bearer ".concat(sessionToken) : undefined,
    [constants.SUPPORTED_HEADERS.X_APPLICATION_ID]: window.app.applicationIdentifier,
    [constants.SUPPORTED_HEADERS.X_CUSTOM_VIEW_ID]: window.app.customViewId,
    [constants.SUPPORTED_HEADERS.X_TEAM_ID]: applicationShellConnectors.selectTeamIdFromStorage()
  });
};
const sdkMiddleware = sdk.createMiddleware({
  getCorrelationId: patchedGetCorrelationId,
  getProjectKey: applicationShellConnectors.selectProjectKeyFromUrl,
  getAdditionalHeaders
});

/**
 * @deprecated: This function should not be used directly anymore.
 */
const applyDefaultMiddlewares = function () {
  for (var _len = arguments.length, middlewares = new Array(_len), _key = 0; _key < _len; _key++) {
    middlewares[_key] = arguments[_key];
  }
  return toolkit.applyMiddleware(...middlewares, thunk__default["default"], loggerMiddleware);
};
const createInternalReducer = function () {
  let injectedReducers = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  let preloadedState = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
  // not providing preloadedStateReducers to combineReducers will cause an error
  // when preloadedState contains keys other than requestsInFlight and
  // notifications:
  // Unexpected key "new Key" found in preloadedState argument passed to createStore. Expected to find one of the known reducer keys instead: "requestsInFlight", "notifications". Unexpected keys will be ignored.
  // https://redux.js.org/api/createstore#createstorereducer-preloadedstate-enhancer
  const preloadedStateReducers = mapValues__default["default"](preloadedState, value => () => value);

  // NOTE: since we don't know in advance which reducers will be injected,
  // we pass an `unknown` type to express this uncertainty and make the compiler happy.
  return toolkit.combineReducers(_objectSpread(_objectSpread({
    requestsInFlight: requestsInFlightReducer,
    notifications: notifications.reducer
  }, injectedReducers), preloadedStateReducers));
};

// We use a factory as it's more practicable for tests
// The application can import the configured store (the default export)
const createReduxStore = function () {
  let preloadedState = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  let additionalMiddlewares = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : [];
  const store = toolkit.configureStore({
    preloadedState,
    reducer: createInternalReducer(undefined, preloadedState),
    devTools: {
      actionsDenylist: [constants.SHOW_LOADING, constants.HIDE_LOADING]
    },
    middleware: getDefaultMiddleware => [...additionalMiddlewares, hideNotificationsMiddleware, notifications.middleware, sdkMiddleware, ...getDefaultMiddleware({
      // https://redux-toolkit.js.org/usage/usage-guide#working-with-non-serializable-data
      serializableCheck: false,
      // This default check is logging warnings to console for some MC FE tests, probably
      // due to lower resources when running tests on CI.
      // We don't consider this check to be very valuable in this context so we decided
      // to turn it off.
      // https://redux-toolkit.js.org/api/immutabilityMiddleware
      immutableCheck: false
    }), loggerMiddleware]
  });
  const enhancedStore = store;

  // Enable reducers to be injected on runtime (see `<InjectReducer>`)
  enhancedStore.injectedReducers = {};
  enhancedStore.injectReducers = _ref3 => {
    let id = _ref3.id,
      reducers = _ref3.reducers;
    const hasReducerBeenInjected = _Reflect$has__default["default"](enhancedStore.injectedReducers, id) && enhancedStore.injectedReducers[id] === reducers;
    if (!hasReducerBeenInjected) {
      // Keep track of the reducer by id, so we can check if it's been injected aleady
      enhancedStore.injectedReducers[id] = reducers;
      // ...when we create the new reducer though, we spread the reducers from each namespace
      enhancedStore.replaceReducer(createInternalReducer(mergeObjectValues(enhancedStore.injectedReducers)));
    }
  };
  enhancedStore.removeReducers = _ref4 => {
    let id = _ref4.id;
    delete enhancedStore.injectedReducers[id];
    enhancedStore.replaceReducer(createInternalReducer(mergeObjectValues(enhancedStore.injectedReducers)));
  };
  return enhancedStore;
};
var internalReduxStore = createReduxStore();

const RouteCatchAll = /*#__PURE__*/react.lazy(() => Promise.resolve().then(function () { return require('./route-catch-all-373c69c6.cjs.prod.js' /* webpackChunkName: "route-catch-all" */); }));

exports.RouteCatchAll = RouteCatchAll;
exports.applyDefaultMiddlewares = applyDefaultMiddlewares;
exports.createReduxStore = createReduxStore;
exports.internalReduxStore = internalReduxStore;
