// @commercetools-frontend/application-shell/dist/version-tracker-e89a6e6b.esm.js
import { useEffect, version } from 'react';
import { useApplicationContext } from '@commercetools-frontend/application-shell-connectors';
import { actions, useAsyncDispatch } from '@commercetools-frontend/sdk';
import { d as applicationShellVersion } from './index-2b55a829.esm.js';
import _JSON$stringify from '@babel/runtime-corejs3/core-js-stable/json/stringify';
import { MC_API_PROXY_TARGETS } from '@commercetools-frontend/constants';
import 'prop-types';
import 'react-router-dom';
import './index-89e5711a.esm.js';
import '@babel/runtime-corejs3/core-js-stable/object/keys';
import '@babel/runtime-corejs3/core-js-stable/object/get-own-property-symbols';
import '@babel/runtime-corejs3/core-js-stable/instance/filter';
import '@babel/runtime-corejs3/core-js-stable/object/get-own-property-descriptor';
import '@babel/runtime-corejs3/core-js-stable/instance/for-each';
import '@babel/runtime-corejs3/core-js-stable/object/get-own-property-descriptors';
import '@babel/runtime-corejs3/core-js-stable/object/define-properties';
import '@babel/runtime-corejs3/core-js-stable/object/define-property';
import '@babel/runtime-corejs3/helpers/defineProperty';
import '@babel/runtime-corejs3/helpers/slicedToArray';
import '@babel/runtime-corejs3/core-js-stable/instance/reduce';
import '@babel/runtime-corejs3/core-js-stable/object/entries';
import '@babel/runtime-corejs3/core-js-stable/reflect/has';
import '@reduxjs/toolkit';
import 'lodash/mapValues';
import 'omit-empty-es';
import 'redux-thunk';
import '@commercetools-frontend/notifications';
import '@babel/runtime-corejs3/core-js-stable/instance/index-of';
import '@babel/runtime-corejs3/core-js-stable/instance/slice';
import '@babel/runtime-corejs3/core-js-stable/instance/includes';
import '@commercetools-frontend/sentry';
import '@commercetools-frontend/react-notifications';
import '@babel/runtime-corejs3/core-js-stable/instance/starts-with';
import 'redux-logger';
import '@emotion/styled/base';
import '@babel/runtime-corejs3/core-js-stable/instance/find';
import '@babel/runtime-corejs3/core-js-stable/array/is-array';
import '@emotion/react';
import '@commercetools-frontend/application-components';
import '@commercetools-frontend/i18n';
import './oidc-da8526e3.esm.js';
import '@babel/runtime-corejs3/core-js-stable/instance/concat';
import '@babel/runtime-corejs3/core-js-stable/instance/map';
import '@babel/runtime-corejs3/core-js-stable/url';
import '@commercetools-uikit/design-system';
import '@commercetools-uikit/spacings';
import 'react-intl';
import '@commercetools-uikit/flat-button';
import '@commercetools-uikit/icons';
import './use-is-served-by-proxy-77984220.esm.js';
import '@emotion/react/jsx-runtime';
import '@babel/runtime-corejs3/helpers/taggedTemplateLiteral';
import '@babel/runtime-corejs3/helpers/objectDestructuringEmpty';
import '@babel/runtime-corejs3/helpers/extends';
import 'memoize-one';
import 'react-select';
import '@commercetools-uikit/accessible-hidden';
import '@commercetools-uikit/select-input';
import '@commercetools-frontend/assets/images/ct-logo.svg';
import '@commercetools-uikit/loading-spinner';
import '@commercetools-frontend/browser-history';
import '@commercetools-frontend/l10n';
import '@babel/runtime-corejs3/core-js-stable/reflect/construct';
import '@babel/runtime-corejs3/helpers/classCallCheck';
import '@babel/runtime-corejs3/helpers/createClass';
import '@babel/runtime-corejs3/helpers/inherits';
import '@babel/runtime-corejs3/helpers/possibleConstructorReturn';
import '@babel/runtime-corejs3/helpers/getPrototypeOf';
import '@babel/runtime-corejs3/core-js-stable/object/from-entries';
import '@babel/runtime-corejs3/core-js-stable/instance/flags';
import '@apollo/client/react';
import '@flopflip/combine-adapters';
import '@flopflip/http-adapter';
import '@flopflip/launchdarkly-adapter';
import '@flopflip/react-broadcast';
import 'react-redux';
import 'lodash/upperFirst';
import '@commercetools-uikit/design-system/materials/resets.css';
import '@commercetools-frontend/application-config/ssr';
import '@commercetools-frontend/permissions';
import '@babel/runtime-corejs3/core-js-stable/instance/some';
import '@commercetools-frontend/actions-global';

// eslint-disable-next-line import/prefer-default-export
const pushDependencyVersionCounter = _ref => {
  let payload = _ref.payload;
  return actions.post({
    uri: '/metrics/counters',
    mcApiProxyTarget: MC_API_PROXY_TARGETS.MC_METRICS,
    payload: _JSON$stringify(payload)
  });
};

const createVersionMetric = _ref => {
  let applicationName = _ref.applicationName;
  return [{
    metricName: 'npm_dependency_versions',
    metricLabels: {
      application: applicationName,
      package_name: '@commercetools-frontend/application-shell',
      package_version: applicationShellVersion
    }
  }, {
    metricName: 'npm_dependency_versions',
    metricLabels: {
      application: applicationName,
      package_name: 'react',
      package_version: version
    }
  }];
};
const VersionTracker = () => {
  const applicationName = useApplicationContext(context => context.environment.applicationName);
  const dispatch = useAsyncDispatch();
  useEffect(() => {
    dispatch(pushDependencyVersionCounter({
      payload: createVersionMetric({
        applicationName
      })
    })).catch(() => {
      // Error is ignored under the assumption that page is being
      // reloaded whilst the request was being sent or network request was interrupted
    });
  }, [applicationName, dispatch]);
  return null;
};
VersionTracker.displayName = 'VersionTracker';

export { VersionTracker as default };
