// @commercetools-uikit/localized-text-field/dist/commercetools-uikit-localized-text-field.esm.js
import _defineProperty from '@babel/runtime-corejs3/helpers/esm/defineProperty';
import _classCallCheck from '@babel/runtime-corejs3/helpers/esm/classCallCheck';
import _createClass from '@babel/runtime-corejs3/helpers/esm/createClass';
import _inherits from '@babel/runtime-corejs3/helpers/esm/inherits';
import _possibleConstructorReturn from '@babel/runtime-corejs3/helpers/esm/possibleConstructorReturn';
import _getPrototypeOf from '@babel/runtime-corejs3/helpers/esm/getPrototypeOf';
import _pt from 'prop-types';
import _someInstanceProperty from '@babel/runtime-corejs3/core-js-stable/instance/some';
import _Object$values from '@babel/runtime-corejs3/core-js-stable/object/values';
import _Reflect$construct from '@babel/runtime-corejs3/core-js-stable/reflect/construct';
import _Object$keys from '@babel/runtime-corejs3/core-js-stable/object/keys';
import _Object$getOwnPropertySymbols from '@babel/runtime-corejs3/core-js-stable/object/get-own-property-symbols';
import _filterInstanceProperty from '@babel/runtime-corejs3/core-js-stable/instance/filter';
import _Object$getOwnPropertyDescriptor from '@babel/runtime-corejs3/core-js-stable/object/get-own-property-descriptor';
import _forEachInstanceProperty from '@babel/runtime-corejs3/core-js-stable/instance/for-each';
import _Object$getOwnPropertyDescriptors from '@babel/runtime-corejs3/core-js-stable/object/get-own-property-descriptors';
import _Object$defineProperties from '@babel/runtime-corejs3/core-js-stable/object/define-properties';
import _Object$defineProperty from '@babel/runtime-corejs3/core-js-stable/object/define-property';
import { isValidElement, Component } from 'react';
import { createSequentialId, warning, filterDataAttributes, getFieldId } from '@commercetools-uikit/utils';
import Constraints from '@commercetools-uikit/constraints';
import Spacings from '@commercetools-uikit/spacings';
import FieldLabel from '@commercetools-uikit/field-label';
import LocalizedTextInput from '@commercetools-uikit/localized-text-input';
import FieldErrors from '@commercetools-uikit/field-errors';
import { jsx, jsxs } from '@emotion/react/jsx-runtime';

function ownKeys(object, enumerableOnly) { var keys = _Object$keys(object); if (_Object$getOwnPropertySymbols) { var symbols = _Object$getOwnPropertySymbols(object); enumerableOnly && (symbols = _filterInstanceProperty(symbols).call(symbols, function (sym) { return _Object$getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }
function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var _context2, _context3; var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? _forEachInstanceProperty(_context2 = ownKeys(Object(source), !0)).call(_context2, function (key) { _defineProperty(target, key, source[key]); }) : _Object$getOwnPropertyDescriptors ? _Object$defineProperties(target, _Object$getOwnPropertyDescriptors(source)) : _forEachInstanceProperty(_context3 = ownKeys(Object(source))).call(_context3, function (key) { _Object$defineProperty(target, key, _Object$getOwnPropertyDescriptor(source, key)); }); } return target; }
function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = _Reflect$construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }
function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !_Reflect$construct) return false; if (_Reflect$construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(_Reflect$construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }
const sequentialId = createSequentialId('localized-text-field-');
const sequentialErrorsId = createSequentialId('localized-text-field-error-')();
const hasErrors = errors => {
  var _context;
  return errors && _someInstanceProperty(_context = _Object$values(errors)).call(_context, Boolean);
};
let LocalizedTextField = /*#__PURE__*/function (_Component) {
  _inherits(LocalizedTextField, _Component);
  var _super = _createSuper(LocalizedTextField);
  function LocalizedTextField() {
    var _this;
    _classCallCheck(this, LocalizedTextField);
    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }
    _this = _super.call(this, ...args);
    _this.state = {
      // We generate an id in case no id is provided by the parent to attach the
      // label to the input component.
      id: _this.props.id
    };
    return _this;
  }
  _createClass(LocalizedTextField, [{
    key: "render",
    value: function render() {
      const hasError = this.props.touched && hasErrors(this.props.errors);
      if (this.props.hintIcon) {
        process.env.NODE_ENV !== "production" ? warning(typeof this.props.hint === 'string' || /*#__PURE__*/isValidElement(this.props.hint), 'LocalizedTextField: `hint` is required to be string or ReactNode if hintIcon is present') : void 0;
      }
      if (this.props.hideLanguageExpansionControls) {
        process.env.NODE_ENV !== "production" ? warning(typeof this.props.defaultExpandLanguages !== 'boolean', 'LocalizedTextField: "defaultExpandLanguages" does not have any effect when "hideLanguageExpansionControls" is set.') : void 0;
      }
      return jsx(Constraints.Horizontal, {
        max: this.props.horizontalConstraint,
        children: jsxs(Spacings.Stack, {
          scale: "xs",
          children: [jsx(FieldLabel, {
            title: this.props.title,
            hint: this.props.hint,
            description: this.props.description,
            onInfoButtonClick: this.props.onInfoButtonClick,
            hintIcon: this.props.hintIcon,
            badge: this.props.badge,
            hasRequiredIndicator: this.props.isRequired,
            htmlFor: this.state.id
          }), jsx(LocalizedTextInput, _objectSpread(_objectSpread({
            autoComplete: this.props.autoComplete,
            id: this.state.id,
            name: this.props.name,
            value: this.props.value,
            onChange: this.props.onChange,
            selectedLanguage: this.props.selectedLanguage,
            onBlur: this.props.onBlur,
            onFocus: this.props.onFocus,
            hideLanguageExpansionControls: this.props.hideLanguageExpansionControls,
            defaultExpandLanguages: this.props.defaultExpandLanguages,
            isAutofocussed: this.props.isAutofocussed,
            isDisabled: this.props.isDisabled,
            isReadOnly: this.props.isReadOnly,
            errors: this.props.errorsByLanguage,
            hasError: hasError,
            placeholder: this.props.placeholder,
            horizontalConstraint: "scale"
          }, filterDataAttributes(this.props)), {}, {
            "aria-invalid": hasError,
            "aria-errormessage": sequentialErrorsId
          })), jsx(FieldErrors, {
            id: sequentialErrorsId,
            errors: this.props.errors,
            isVisible: hasError,
            renderError: this.props.renderError
          })]
        })
      });
    }
  }], [{
    key: "toFieldErrors",
    value:
    /**
     * Use this function to convert the Formik `errors` object type to
     * our custom field errors type.
     * This is primarly useful when using TypeScript.
     */
    function toFieldErrors(errors) {
      return errors;
    }
  }]);
  return LocalizedTextField;
}(Component);
LocalizedTextField.displayName = 'LocalizedTextField';
LocalizedTextField.defaultProps = {
  horizontalConstraint: 'scale'
};
LocalizedTextField.getDerivedStateFromProps = (props, state) => ({
  id: getFieldId(props, state, sequentialId)
});
LocalizedTextField.propTypes = process.env.NODE_ENV !== "production" ? {
  id: _pt.string,
  horizontalConstraint: _pt.oneOf([6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 'scale', 'auto']),
  errors: _pt.objectOf(_pt.bool),
  renderError: _pt.func,
  isRequired: _pt.bool,
  touched: _pt.bool,
  autoComplete: _pt.string,
  name: _pt.string,
  value: _pt.objectOf(_pt.string).isRequired,
  onChange: _pt.func,
  selectedLanguage: _pt.string.isRequired,
  onBlur: _pt.func,
  onFocus: _pt.func,
  hideLanguageExpansionControls: _pt.bool,
  defaultExpandLanguages: _pt.bool,
  isAutofocussed: _pt.bool,
  isDisabled: _pt.bool,
  isReadOnly: _pt.bool,
  placeholder: _pt.objectOf(_pt.string),
  errorsByLanguage: _pt.objectOf(_pt.string),
  title: _pt.node.isRequired,
  hint: _pt.node,
  description: _pt.node,
  onInfoButtonClick: _pt.func,
  hintIcon: _pt.element
} : {};
var LocalizedTextField$1 = LocalizedTextField;

// NOTE: This string will be replaced on build time with the package version.
var version = "16.11.0";

export { LocalizedTextField$1 as default, version };
