// @commercetools-uikit/async-creatable-select-input/dist/commercetools-uikit-async-creatable-select-input.cjs.prod.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _defineProperty = require('@babel/runtime-corejs3/helpers/defineProperty');
require('prop-types');
require('@babel/runtime-corejs3/core-js-stable/array/is-array');
var _Object$keys = require('@babel/runtime-corejs3/core-js-stable/object/keys');
var _Object$getOwnPropertySymbols = require('@babel/runtime-corejs3/core-js-stable/object/get-own-property-symbols');
var _filterInstanceProperty = require('@babel/runtime-corejs3/core-js-stable/instance/filter');
var _Object$getOwnPropertyDescriptor = require('@babel/runtime-corejs3/core-js-stable/object/get-own-property-descriptor');
var _forEachInstanceProperty = require('@babel/runtime-corejs3/core-js-stable/instance/for-each');
var _Object$getOwnPropertyDescriptors = require('@babel/runtime-corejs3/core-js-stable/object/get-own-property-descriptors');
var _Object$defineProperties = require('@babel/runtime-corejs3/core-js-stable/object/define-properties');
var _Object$defineProperty = require('@babel/runtime-corejs3/core-js-stable/object/define-property');
var reactIntl = require('react-intl');
var isEmpty = require('lodash/isEmpty');
var reactSelect = require('react-select');
var AsyncCreatableSelect = require('react-select/async-creatable');
var Constraints = require('@commercetools-uikit/constraints');
var LoadingSpinner = require('@commercetools-uikit/loading-spinner');
var selectUtils = require('@commercetools-uikit/select-utils');
var utils = require('@commercetools-uikit/utils');
var jsxRuntime = require('@emotion/react/jsx-runtime');

function _interopDefault (e) { return e && e.__esModule ? e : { 'default': e }; }

var _Object$keys__default = /*#__PURE__*/_interopDefault(_Object$keys);
var _Object$getOwnPropertySymbols__default = /*#__PURE__*/_interopDefault(_Object$getOwnPropertySymbols);
var _filterInstanceProperty__default = /*#__PURE__*/_interopDefault(_filterInstanceProperty);
var _Object$getOwnPropertyDescriptor__default = /*#__PURE__*/_interopDefault(_Object$getOwnPropertyDescriptor);
var _forEachInstanceProperty__default = /*#__PURE__*/_interopDefault(_forEachInstanceProperty);
var _Object$getOwnPropertyDescriptors__default = /*#__PURE__*/_interopDefault(_Object$getOwnPropertyDescriptors);
var _Object$defineProperties__default = /*#__PURE__*/_interopDefault(_Object$defineProperties);
var _Object$defineProperty__default = /*#__PURE__*/_interopDefault(_Object$defineProperty);
var isEmpty__default = /*#__PURE__*/_interopDefault(isEmpty);
var AsyncCreatableSelect__default = /*#__PURE__*/_interopDefault(AsyncCreatableSelect);
var Constraints__default = /*#__PURE__*/_interopDefault(Constraints);
var LoadingSpinner__default = /*#__PURE__*/_interopDefault(LoadingSpinner);

function ownKeys(object, enumerableOnly) { var keys = _Object$keys__default["default"](object); if (_Object$getOwnPropertySymbols__default["default"]) { var symbols = _Object$getOwnPropertySymbols__default["default"](object); enumerableOnly && (symbols = _filterInstanceProperty__default["default"](symbols).call(symbols, function (sym) { return _Object$getOwnPropertyDescriptor__default["default"](object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }
function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var _context, _context2; var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? _forEachInstanceProperty__default["default"](_context = ownKeys(Object(source), !0)).call(_context, function (key) { _defineProperty(target, key, source[key]); }) : _Object$getOwnPropertyDescriptors__default["default"] ? _Object$defineProperties__default["default"](target, _Object$getOwnPropertyDescriptors__default["default"](source)) : _forEachInstanceProperty__default["default"](_context2 = ownKeys(Object(source))).call(_context2, function (key) { _Object$defineProperty__default["default"](target, key, _Object$getOwnPropertyDescriptor__default["default"](source, key)); }); } return target; }
const LoadingIndicator = () => jsxRuntime.jsx(LoadingSpinner__default["default"], {
  scale: "s"
});
LoadingIndicator.displayName = 'LoadingIndicator';
const customizedComponents = {
  DropdownIndicator: selectUtils.DropdownIndicator,
  ClearIndicator: selectUtils.ClearIndicator,
  LoadingIndicator,
  MultiValueRemove: selectUtils.TagRemove
};
const defaultProps = {
  // Using "null" will ensure that the currently selected value disappears in
  // case "undefined" gets passed as the next value
  value: null,
  isSearchable: true,
  menuPortalZIndex: 1
};
const AsyncCreatableSelectInput = props => {
  const intl = reactIntl.useIntl();
  const placeholder = props.placeholder || intl.formatMessage(selectUtils.messages.placeholder);
  if (!props.isReadOnly) ;
  if (props.isMulti) ;
  selectUtils.warnIfMenuPortalPropsAreMissing({
    menuPortalZIndex: props.menuPortalZIndex,
    menuPortalTarget: props.menuPortalTarget,
    componentName: 'AsyncCreatableSelectInput'
  });
  return jsxRuntime.jsx(Constraints__default["default"].Horizontal, {
    max: props.horizontalConstraint,
    children: jsxRuntime.jsx("div", _objectSpread(_objectSpread({}, utils.filterDataAttributes(props)), {}, {
      children: jsxRuntime.jsx(AsyncCreatableSelect__default["default"], {
        "aria-label": props['aria-label'],
        "aria-labelledby": props['aria-labelledby'],
        "aria-invalid": props['aria-invalid'],
        "aria-errormessage": props['aria-errormessage'],
        autoFocus: props.isAutofocussed,
        backspaceRemovesValue: props.isReadOnly ? false : props.backspaceRemovesValue,
        components: _objectSpread(_objectSpread(_objectSpread(_objectSpread({}, customizedComponents), props.iconLeft && !props.isMulti ? selectUtils.customComponentsWithIcons : {}), props.isReadOnly ? {
          Input: ownProps => jsxRuntime.jsx(reactSelect.components.Input, _objectSpread(_objectSpread({}, ownProps), {}, {
            readOnly: true
          }))
        } : {}), props.components),
        menuIsOpen: props.isReadOnly ? false : undefined,
        styles: selectUtils.createSelectStyles({
          hasWarning: props.hasWarning,
          hasError: props.hasError,
          showOptionGroupDivider: props.showOptionGroupDivider,
          menuPortalZIndex: props.menuPortalZIndex,
          isDisabled: props.isDisabled,
          isReadOnly: props.isReadOnly,
          iconLeft: props.iconLeft,
          isMulti: props.isMulti,
          hasValue: !isEmpty__default["default"](props.value),
          horizontalConstraint: props.horizontalConstraint
        }),
        filterOption: props.filterOption
        // react-select uses "id" (for the container) and "inputId" (for the input),
        // but we use "id" (for the input) and "containerId" (for the container)
        // instead.
        // This makes it easier to less confusing to use with <label />s.
        ,
        id: props.containerId,
        inputId: props.id,
        inputValue: props.inputValue,
        isClearable: props.isReadOnly ? false : props.isClearable,
        isDisabled: props.isDisabled,
        isOptionDisabled: props.isOptionDisabled,
        isMulti: props.isMulti,
        isSearchable: props.isSearchable,
        maxMenuHeight: props.maxMenuHeight,
        menuPortalTarget: props.menuPortalTarget,
        menuShouldBlockScroll: props.menuShouldBlockScroll,
        closeMenuOnScroll: props.closeMenuOnSelect,
        name: props.name,
        noOptionsMessage: props.noOptionsMessage || (_ref => {
          let inputValue = _ref.inputValue;
          return inputValue === '' ? intl.formatMessage(selectUtils.messages.noOptionsMessageWithoutInputValue) : intl.formatMessage(selectUtils.messages.noOptionsMessageWithInputValue, {
            inputValue
          });
        }),
        onBlur: props.onBlur ? () => {
          const event = {
            target: {
              id: props.id,
              name: (() => {
                if (!props.name) return undefined;
                if (!props.isMulti) return props.name;
                // We append the ".0" to make Formik set the touched
                // state as an array instead of a boolean only.
                // Otherwise the shapes would clash on submission, as
                // Formik will create an array on submission anyways.
                return "".concat(props.name, ".0");
              })()
            },
            persist: () => {}
          };
          props.onBlur && props.onBlur(event);
        } : undefined,
        onChange: (value, info) => {
          var _props$onChange;
          // wrapping breaking changes made in react-select v3
          let newValue = value;
          if (props.isMulti && !newValue) {
            newValue = [];
          }
          (_props$onChange = props.onChange) === null || _props$onChange === void 0 ? void 0 : _props$onChange.call(props, {
            target: {
              id: props.id,
              name: props.name,
              value: newValue
            },
            persist: () => {}
          }, info);
        },
        onFocus: props.onFocus,
        onInputChange: props.onInputChange,
        placeholder: placeholder,
        tabIndex: props.tabIndex,
        tabSelectsValue: props.tabSelectsValue,
        value: props.value
        // Async react-select props
        ,
        defaultOptions: props.defaultOptions,
        loadOptions: props.loadOptions,
        cacheOptions: props.cacheOptions
        // Creatable props
        ,
        allowCreateWhileLoading: props.allowCreateWhileLoading,
        formatCreateLabel: props.formatCreateLabel || (inputValue => intl.formatMessage(selectUtils.messages.createLabel, {
          inputValue
        })),
        isValidNewOption: props.isValidNewOption,
        getNewOptionData: props.getNewOptionData,
        onCreateOption: props.onCreateOption,
        createOptionPosition: props.createOptionPosition
        // @ts-ignore
        ,
        iconLeft: props.iconLeft
      })
    }))
  });
};
AsyncCreatableSelectInput.propTypes = {};
AsyncCreatableSelectInput.displayName = 'AsyncCreatableSelectInput';
AsyncCreatableSelectInput.defaultProps = defaultProps;

/**
 * Expose static helper methods.
 */

// Formik will set the field to an array on submission, so we always have to
// deal with an array. The touched state ends up being an empty array in case
// values were removed only. So we have to treat any array we receive as
// a signal of the field having been touched.
AsyncCreatableSelectInput.isTouched = touched => Boolean(touched);

/**
 * Expose react-select components for customization purposes.
 */

// custom
AsyncCreatableSelectInput.ClearIndicator = customizedComponents.ClearIndicator;
AsyncCreatableSelectInput.Control = reactSelect.components.Control;
AsyncCreatableSelectInput.CrossIcon = reactSelect.components.CrossIcon;
AsyncCreatableSelectInput.DownChevron = reactSelect.components.DownChevron;
// custom
AsyncCreatableSelectInput.DropdownIndicator = customizedComponents.DropdownIndicator;
AsyncCreatableSelectInput.Group = reactSelect.components.Group;
AsyncCreatableSelectInput.GroupHeading = reactSelect.components.GroupHeading;
AsyncCreatableSelectInput.IndicatorSeparator = reactSelect.components.IndicatorSeparator;
AsyncCreatableSelectInput.IndicatorsContainer = reactSelect.components.IndicatorsContainer;
AsyncCreatableSelectInput.Input = reactSelect.components.Input;
// custom
AsyncCreatableSelectInput.LoadingIndicator = customizedComponents.LoadingIndicator;
AsyncCreatableSelectInput.LoadingMessage = reactSelect.components.LoadingMessage;
AsyncCreatableSelectInput.Menu = reactSelect.components.Menu;
AsyncCreatableSelectInput.MenuList = reactSelect.components.MenuList;
AsyncCreatableSelectInput.MenuPortal = reactSelect.components.MenuPortal;
AsyncCreatableSelectInput.MultiValue = reactSelect.components.MultiValue;
AsyncCreatableSelectInput.MultiValueContainer = reactSelect.components.MultiValueContainer;
AsyncCreatableSelectInput.MultiValueLabel = reactSelect.components.MultiValueLabel;
// custom
AsyncCreatableSelectInput.MultiValueRemove = customizedComponents.MultiValueRemove;
AsyncCreatableSelectInput.NoOptionsMessage = reactSelect.components.NoOptionsMessage;
AsyncCreatableSelectInput.Option = reactSelect.components.Option;
AsyncCreatableSelectInput.Placeholder = reactSelect.components.Placeholder;
AsyncCreatableSelectInput.SelectContainer = reactSelect.components.SelectContainer;
AsyncCreatableSelectInput.SingleValue = reactSelect.components.SingleValue;
AsyncCreatableSelectInput.ValueContainer = reactSelect.components.ValueContainer;
var AsyncCreatableSelectInput$1 = AsyncCreatableSelectInput;

// NOTE: This string will be replaced on build time with the package version.
var version = "16.11.0";

exports["default"] = AsyncCreatableSelectInput$1;
exports.version = version;
