// @commercetools-frontend/ui-kit/dist/commercetools-frontend-ui-kit.cjs.prod.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var i18n = require('@commercetools-uikit/i18n');
var buttons = require('@commercetools-uikit/buttons');
var fields = require('@commercetools-uikit/fields');
var icons = require('@commercetools-uikit/icons');
var inputs = require('@commercetools-uikit/inputs');
var accessibleHidden = require('@commercetools-uikit/accessible-hidden');
var avatar = require('@commercetools-uikit/avatar');
var card = require('@commercetools-uikit/card');
var link = require('@commercetools-uikit/link');
var collapsible = require('@commercetools-uikit/collapsible');
var collapsibleMotion = require('@commercetools-uikit/collapsible-motion');
var collapsiblePanel = require('@commercetools-uikit/collapsible-panel');
var constraints = require('@commercetools-uikit/constraints');
var notifications = require('@commercetools-uikit/notifications');
var primaryActionDropdown = require('@commercetools-uikit/primary-action-dropdown');
var fieldLabel = require('@commercetools-uikit/field-label');
var fieldErrors = require('@commercetools-uikit/field-errors');
var grid = require('@commercetools-uikit/grid');
var label = require('@commercetools-uikit/label');
var loadingSpinner = require('@commercetools-uikit/loading-spinner');
var messages = require('@commercetools-uikit/messages');
var pagination = require('@commercetools-uikit/pagination');
var spacings = require('@commercetools-uikit/spacings');
var stamp = require('@commercetools-uikit/stamp');
var dataTable = require('@commercetools-uikit/data-table');
var dataTableManager = require('@commercetools-uikit/data-table-manager');
var tag = require('@commercetools-uikit/tag');
var tooltip = require('@commercetools-uikit/tooltip');
var text = require('@commercetools-uikit/text');
var viewSwitcher = require('@commercetools-uikit/view-switcher');
var hooks = require('@commercetools-uikit/hooks');
var designSystem = require('@commercetools-uikit/design-system');

function _interopDefault (e) { return e && e.__esModule ? e : { 'default': e }; }

function _interopNamespace(e) {
  if (e && e.__esModule) return e;
  var n = Object.create(null);
  if (e) {
    Object.keys(e).forEach(function (k) {
      if (k !== 'default') {
        var d = Object.getOwnPropertyDescriptor(e, k);
        Object.defineProperty(n, k, d.get ? d : {
          enumerable: true,
          get: function () { return e[k]; }
        });
      }
    });
  }
  n["default"] = e;
  return Object.freeze(n);
}

var i18n__namespace = /*#__PURE__*/_interopNamespace(i18n);
var accessibleHidden__default = /*#__PURE__*/_interopDefault(accessibleHidden);
var avatar__default = /*#__PURE__*/_interopDefault(avatar);
var card__default = /*#__PURE__*/_interopDefault(card);
var link__default = /*#__PURE__*/_interopDefault(link);
var collapsible__default = /*#__PURE__*/_interopDefault(collapsible);
var collapsibleMotion__default = /*#__PURE__*/_interopDefault(collapsibleMotion);
var collapsiblePanel__default = /*#__PURE__*/_interopDefault(collapsiblePanel);
var constraints__default = /*#__PURE__*/_interopDefault(constraints);
var primaryActionDropdown__default = /*#__PURE__*/_interopDefault(primaryActionDropdown);
var fieldLabel__default = /*#__PURE__*/_interopDefault(fieldLabel);
var fieldErrors__default = /*#__PURE__*/_interopDefault(fieldErrors);
var grid__default = /*#__PURE__*/_interopDefault(grid);
var label__default = /*#__PURE__*/_interopDefault(label);
var loadingSpinner__default = /*#__PURE__*/_interopDefault(loadingSpinner);
var spacings__default = /*#__PURE__*/_interopDefault(spacings);
var stamp__default = /*#__PURE__*/_interopDefault(stamp);
var dataTable__default = /*#__PURE__*/_interopDefault(dataTable);
var dataTableManager__default = /*#__PURE__*/_interopDefault(dataTableManager);
var tooltip__default = /*#__PURE__*/_interopDefault(tooltip);
var text__default = /*#__PURE__*/_interopDefault(text);
var viewSwitcher__default = /*#__PURE__*/_interopDefault(viewSwitcher);

// NOTE: This string will be replaced on build time with the package version.
var version = "16.11.0";

exports.i18n = i18n__namespace;
Object.defineProperty(exports, 'AccessibleHidden', {
  enumerable: true,
  get: function () { return accessibleHidden__default["default"]; }
});
Object.defineProperty(exports, 'Avatar', {
  enumerable: true,
  get: function () { return avatar__default["default"]; }
});
Object.defineProperty(exports, 'Card', {
  enumerable: true,
  get: function () { return card__default["default"]; }
});
Object.defineProperty(exports, 'Link', {
  enumerable: true,
  get: function () { return link__default["default"]; }
});
Object.defineProperty(exports, 'Collapsible', {
  enumerable: true,
  get: function () { return collapsible__default["default"]; }
});
Object.defineProperty(exports, 'CollapsibleMotion', {
  enumerable: true,
  get: function () { return collapsibleMotion__default["default"]; }
});
Object.defineProperty(exports, 'CollapsiblePanel', {
  enumerable: true,
  get: function () { return collapsiblePanel__default["default"]; }
});
Object.defineProperty(exports, 'Constraints', {
  enumerable: true,
  get: function () { return constraints__default["default"]; }
});
Object.defineProperty(exports, 'ContentNotification', {
  enumerable: true,
  get: function () { return notifications.ContentNotification; }
});
Object.defineProperty(exports, 'PrimaryActionDropdown', {
  enumerable: true,
  get: function () { return primaryActionDropdown__default["default"]; }
});
Object.defineProperty(exports, 'PrimaryActionDropdownOption', {
  enumerable: true,
  get: function () { return primaryActionDropdown.Option; }
});
Object.defineProperty(exports, 'FieldLabel', {
  enumerable: true,
  get: function () { return fieldLabel__default["default"]; }
});
Object.defineProperty(exports, 'FieldErrors', {
  enumerable: true,
  get: function () { return fieldErrors__default["default"]; }
});
Object.defineProperty(exports, 'Grid', {
  enumerable: true,
  get: function () { return grid__default["default"]; }
});
Object.defineProperty(exports, 'Label', {
  enumerable: true,
  get: function () { return label__default["default"]; }
});
Object.defineProperty(exports, 'LoadingSpinner', {
  enumerable: true,
  get: function () { return loadingSpinner__default["default"]; }
});
Object.defineProperty(exports, 'ErrorMessage', {
  enumerable: true,
  get: function () { return messages.ErrorMessage; }
});
Object.defineProperty(exports, 'WarningMessage', {
  enumerable: true,
  get: function () { return messages.WarningMessage; }
});
Object.defineProperty(exports, 'PageNavigator', {
  enumerable: true,
  get: function () { return pagination.PageNavigator; }
});
Object.defineProperty(exports, 'PageSizeSelector', {
  enumerable: true,
  get: function () { return pagination.PageSizeSelector; }
});
Object.defineProperty(exports, 'Pagination', {
  enumerable: true,
  get: function () { return pagination.Pagination; }
});
Object.defineProperty(exports, 'Spacings', {
  enumerable: true,
  get: function () { return spacings__default["default"]; }
});
Object.defineProperty(exports, 'Stamp', {
  enumerable: true,
  get: function () { return stamp__default["default"]; }
});
Object.defineProperty(exports, 'DataTable', {
  enumerable: true,
  get: function () { return dataTable__default["default"]; }
});
Object.defineProperty(exports, 'DataTableManager', {
  enumerable: true,
  get: function () { return dataTableManager__default["default"]; }
});
Object.defineProperty(exports, 'Tag', {
  enumerable: true,
  get: function () { return tag.Tag; }
});
Object.defineProperty(exports, 'TagList', {
  enumerable: true,
  get: function () { return tag.TagList; }
});
Object.defineProperty(exports, 'Tooltip', {
  enumerable: true,
  get: function () { return tooltip__default["default"]; }
});
Object.defineProperty(exports, 'Text', {
  enumerable: true,
  get: function () { return text__default["default"]; }
});
Object.defineProperty(exports, 'ViewSwitcher', {
  enumerable: true,
  get: function () { return viewSwitcher__default["default"]; }
});
Object.defineProperty(exports, 'useDataTableSortingState', {
  enumerable: true,
  get: function () { return hooks.useDataTableSortingState; }
});
Object.defineProperty(exports, 'usePaginationState', {
  enumerable: true,
  get: function () { return hooks.usePaginationState; }
});
Object.defineProperty(exports, 'useRowSelection', {
  enumerable: true,
  get: function () { return hooks.useRowSelection; }
});
Object.defineProperty(exports, 'useSorting', {
  enumerable: true,
  get: function () { return hooks.useSorting; }
});
Object.defineProperty(exports, 'useToggleState', {
  enumerable: true,
  get: function () { return hooks.useToggleState; }
});
Object.defineProperty(exports, 'customProperties', {
  enumerable: true,
  get: function () { return designSystem.customProperties; }
});
Object.defineProperty(exports, 'designTokens', {
  enumerable: true,
  get: function () { return designSystem.designTokens; }
});
exports.version = version;
Object.keys(buttons).forEach(function (k) {
  if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
    enumerable: true,
    get: function () { return buttons[k]; }
  });
});
Object.keys(fields).forEach(function (k) {
  if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
    enumerable: true,
    get: function () { return fields[k]; }
  });
});
Object.keys(icons).forEach(function (k) {
  if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
    enumerable: true,
    get: function () { return icons[k]; }
  });
});
Object.keys(inputs).forEach(function (k) {
  if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
    enumerable: true,
    get: function () { return inputs[k]; }
  });
});
