// @commercetools-uikit/pagination/dist/commercetools-uikit-pagination.cjs.prod.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

require('@emotion/react');
var Spacings = require('@commercetools-uikit/spacings');
var jsxRuntime = require('@emotion/react/jsx-runtime');
var _defineProperty = require('@babel/runtime-corejs3/helpers/defineProperty');
var _slicedToArray = require('@babel/runtime-corejs3/helpers/slicedToArray');
require('prop-types');
var _includesInstanceProperty = require('@babel/runtime-corejs3/core-js-stable/instance/includes');
require('@babel/runtime-corejs3/core-js-stable/instance/concat');
var _mapInstanceProperty = require('@babel/runtime-corejs3/core-js-stable/instance/map');
var _Object$keys = require('@babel/runtime-corejs3/core-js-stable/object/keys');
var _Object$getOwnPropertySymbols = require('@babel/runtime-corejs3/core-js-stable/object/get-own-property-symbols');
var _filterInstanceProperty = require('@babel/runtime-corejs3/core-js-stable/instance/filter');
var _Object$getOwnPropertyDescriptor = require('@babel/runtime-corejs3/core-js-stable/object/get-own-property-descriptor');
var _forEachInstanceProperty = require('@babel/runtime-corejs3/core-js-stable/instance/for-each');
var _Object$getOwnPropertyDescriptors = require('@babel/runtime-corejs3/core-js-stable/object/get-own-property-descriptors');
var _Object$defineProperties = require('@babel/runtime-corejs3/core-js-stable/object/define-properties');
var _Object$defineProperty = require('@babel/runtime-corejs3/core-js-stable/object/define-property');
var react = require('react');
var uniqueId = require('lodash/uniqueId');
var SelectInput = require('@commercetools-uikit/select-input');
var Constraints = require('@commercetools-uikit/constraints');
require('@commercetools-uikit/utils');
var Label = require('@commercetools-uikit/label');
var reactIntl = require('react-intl');
var _valuesInstanceProperty = require('@babel/runtime-corejs3/core-js-stable/instance/values');
var formik = require('formik');
var icons = require('@commercetools-uikit/icons');
var NumberInput = require('@commercetools-uikit/number-input');
var SecondaryIconButton = require('@commercetools-uikit/secondary-icon-button');

function _interopDefault (e) { return e && e.__esModule ? e : { 'default': e }; }

var Spacings__default = /*#__PURE__*/_interopDefault(Spacings);
var _includesInstanceProperty__default = /*#__PURE__*/_interopDefault(_includesInstanceProperty);
var _mapInstanceProperty__default = /*#__PURE__*/_interopDefault(_mapInstanceProperty);
var _Object$keys__default = /*#__PURE__*/_interopDefault(_Object$keys);
var _Object$getOwnPropertySymbols__default = /*#__PURE__*/_interopDefault(_Object$getOwnPropertySymbols);
var _filterInstanceProperty__default = /*#__PURE__*/_interopDefault(_filterInstanceProperty);
var _Object$getOwnPropertyDescriptor__default = /*#__PURE__*/_interopDefault(_Object$getOwnPropertyDescriptor);
var _forEachInstanceProperty__default = /*#__PURE__*/_interopDefault(_forEachInstanceProperty);
var _Object$getOwnPropertyDescriptors__default = /*#__PURE__*/_interopDefault(_Object$getOwnPropertyDescriptors);
var _Object$defineProperties__default = /*#__PURE__*/_interopDefault(_Object$defineProperties);
var _Object$defineProperty__default = /*#__PURE__*/_interopDefault(_Object$defineProperty);
var uniqueId__default = /*#__PURE__*/_interopDefault(uniqueId);
var SelectInput__default = /*#__PURE__*/_interopDefault(SelectInput);
var Constraints__default = /*#__PURE__*/_interopDefault(Constraints);
var Label__default = /*#__PURE__*/_interopDefault(Label);
var _valuesInstanceProperty__default = /*#__PURE__*/_interopDefault(_valuesInstanceProperty);
var NumberInput__default = /*#__PURE__*/_interopDefault(NumberInput);
var SecondaryIconButton__default = /*#__PURE__*/_interopDefault(SecondaryIconButton);

const isValid = (page, totalPages) => {
  if (page > totalPages) return false;
  if (page <= 0) return false;
  return true;
};
const normalizePageValue = (value, totalPages) => {
  if (value < 1) {
    return 1;
  }
  if (value > totalPages) {
    return totalPages;
  }
  return value;
};

var messages$2 = reactIntl.defineMessages({
  page: {
    id: 'UIKit.Pagination.PageNavigator.page',
    description: 'Label for page',
    defaultMessage: 'Page'
  },
  pageCount: {
    id: 'UIKit.Pagination.PageNavigator.pageCount',
    description: 'Label for total pages',
    defaultMessage: 'of {count}'
  },
  previousPageLabel: {
    id: 'UIKit.Pagination.PageNavigator.previousPageLabel',
    description: 'Label for previous page button',
    defaultMessage: 'Previous page'
  },
  nextPageLabel: {
    id: 'UIKit.Pagination.PageNavigator.nextPageLabel',
    description: 'Label for next page button',
    defaultMessage: 'Next page'
  }
});

function ownKeys$1(object, enumerableOnly) { var keys = _Object$keys__default["default"](object); if (_Object$getOwnPropertySymbols__default["default"]) { var symbols = _Object$getOwnPropertySymbols__default["default"](object); enumerableOnly && (symbols = _filterInstanceProperty__default["default"](symbols).call(symbols, function (sym) { return _Object$getOwnPropertyDescriptor__default["default"](object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }
function _objectSpread$1(target) { for (var i = 1; i < arguments.length; i++) { var _context, _context2; var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? _forEachInstanceProperty__default["default"](_context = ownKeys$1(Object(source), !0)).call(_context, function (key) { _defineProperty(target, key, source[key]); }) : _Object$getOwnPropertyDescriptors__default["default"] ? _Object$defineProperties__default["default"](target, _Object$getOwnPropertyDescriptors__default["default"](source)) : _forEachInstanceProperty__default["default"](_context2 = ownKeys$1(Object(source))).call(_context2, function (key) { _Object$defineProperty__default["default"](target, key, _Object$getOwnPropertyDescriptor__default["default"](source, key)); }); } return target; }
const PageNavigator = props => {
  const intl = reactIntl.useIntl();
  const _useState = react.useState(uniqueId__default["default"]('page-number-')),
    _useState2 = _slicedToArray(_useState, 1),
    pageNumberInputId = _useState2[0];
  const paginationForm = formik.useFormik({
    initialValues: {
      page: props.page
    },
    enableReinitialize: true,
    validateOnBlur: true,
    validateOnChange: true,
    validateOnMount: true,
    onSubmit: (values /**, helpers */) => {
      const nextNormalizedValue = Number(normalizePageValue(values.page, props.totalPages));
      props.onPageChange(nextNormalizedValue);
    },
    validate: values => {
      if (!isValid(values.page, props.totalPages)) {
        return {
          page: true
        };
      }
      return {};
    }
  });
  const page = props.page,
    totalPages = props.totalPages;
  const isDisabled = totalPages === 0;
  const isPreviousDisabled = page <= 1;
  const isNextDisabled = page >= totalPages;
  const handlePrevPageChange = react.useCallback(() => {
    const previousPage = _valuesInstanceProperty__default["default"](paginationForm).page - 1;
    if (previousPage >= 1) {
      paginationForm.setFieldValue('page', previousPage, true);
      paginationForm.submitForm();
    }
  }, [paginationForm]);
  const handleNextPageChange = react.useCallback(() => {
    const nextPage = _valuesInstanceProperty__default["default"](paginationForm).page + 1;
    if (nextPage <= totalPages) {
      paginationForm.setFieldValue('page', nextPage, true);
      paginationForm.submitForm();
    }
  }, [paginationForm, totalPages]);
  return jsxRuntime.jsx("form", {
    onSubmit: paginationForm.handleSubmit,
    children: jsxRuntime.jsxs(Spacings__default["default"].Inline, {
      alignItems: "center",
      scale: "s",
      children: [jsxRuntime.jsx(SecondaryIconButton__default["default"], {
        label: intl.formatMessage(messages$2.previousPageLabel),
        onClick: handlePrevPageChange,
        isDisabled: isPreviousDisabled || isDisabled,
        icon: jsxRuntime.jsx(icons.AngleLeftIcon, {})
      }), jsxRuntime.jsx(Label__default["default"], {
        htmlFor: pageNumberInputId,
        intlMessage: messages$2.page
      }), jsxRuntime.jsx("div", {
        children: jsxRuntime.jsx(NumberInput__default["default"], {
          name: "page",
          id: pageNumberInputId,
          value: _valuesInstanceProperty__default["default"](paginationForm).page,
          min: 1,
          max: totalPages,
          onBlur: paginationForm.handleBlur,
          onChange: paginationForm.handleChange,
          isDisabled: isDisabled,
          hasWarning: Boolean(paginationForm.errors.page),
          horizontalConstraint: 2
        })
      }), jsxRuntime.jsx(Label__default["default"], {
        intlMessage: _objectSpread$1(_objectSpread$1({}, messages$2.pageCount), {}, {
          values: {
            count: props.totalPages
          }
        })
      }), jsxRuntime.jsx(SecondaryIconButton__default["default"], {
        label: intl.formatMessage(messages$2.nextPageLabel),
        onClick: handleNextPageChange,
        isDisabled: isNextDisabled || isDisabled,
        icon: jsxRuntime.jsx(icons.AngleRightIcon, {})
      })]
    })
  });
};
PageNavigator.propTypes = {};
PageNavigator.displayName = 'PageNavigator';
var PageNavigator$1 = PageNavigator;

const messages = reactIntl.defineMessages({
  pageSize: {
    id: 'UIKit.Pagination.PageSizeSelector.pageSize',
    description: 'How many items are per page',
    defaultMessage: 'Items per page ({count} items)'
  }
});
var messages$1 = messages;

function ownKeys(object, enumerableOnly) { var keys = _Object$keys__default["default"](object); if (_Object$getOwnPropertySymbols__default["default"]) { var symbols = _Object$getOwnPropertySymbols__default["default"](object); enumerableOnly && (symbols = _filterInstanceProperty__default["default"](symbols).call(symbols, function (sym) { return _Object$getOwnPropertyDescriptor__default["default"](object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }
function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var _context2, _context3; var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? _forEachInstanceProperty__default["default"](_context2 = ownKeys(Object(source), !0)).call(_context2, function (key) { _defineProperty(target, key, source[key]); }) : _Object$getOwnPropertyDescriptors__default["default"] ? _Object$defineProperties__default["default"](target, _Object$getOwnPropertyDescriptors__default["default"](source)) : _forEachInstanceProperty__default["default"](_context3 = ownKeys(Object(source))).call(_context3, function (key) { _Object$defineProperty__default["default"](target, key, _Object$getOwnPropertyDescriptor__default["default"](source, key)); }); } return target; }
const mapRangeToListOfOptions = perPageRange => {
  switch (perPageRange) {
    case 's':
      return [20, 50];
    case 'm':
      return [20, 50, 100];
    case 'l':
      return [200, 500];
    default:
      throw new Error("Invalid page range \"".concat(perPageRange, "\", expected one of \"s,m,l\"."));
  }
};
const PageSizeSelector = props => {
  const _useState = react.useState(uniqueId__default["default"]('per-page-selector-')),
    _useState2 = _slicedToArray(_useState, 1),
    perPageSelectorId = _useState2[0];
  const options = mapRangeToListOfOptions(props.perPageRange);
  _includesInstanceProperty__default["default"](options).call(options, props.perPage);
  const onPerPageChange = props.onPerPageChange;
  const handleSelectPerPage = react.useCallback(event => {
    onPerPageChange(Number(event.target.value));
  }, [onPerPageChange]);
  return jsxRuntime.jsxs(Spacings__default["default"].Inline, {
    alignItems: "center",
    children: [jsxRuntime.jsx(Constraints__default["default"].Horizontal, {
      max: "auto",
      children: jsxRuntime.jsx(SelectInput__default["default"], {
        id: perPageSelectorId,
        isSearchable: false,
        name: "per-page-selector",
        value: props.perPage.toString(),
        options: _mapInstanceProperty__default["default"](options).call(options, option => ({
          value: option.toString(),
          label: option.toString()
        })),
        onChange: handleSelectPerPage
      })
    }), jsxRuntime.jsx(Label__default["default"], {
      htmlFor: perPageSelectorId,
      intlMessage: _objectSpread(_objectSpread({}, messages$1.pageSize), {}, {
        values: {
          count: props.pageItems
        }
      })
    })]
  });
};
PageSizeSelector.propTypes = {};
PageSizeSelector.displayName = 'PageSizeSelector';
const defaultProps$1 = {
  perPage: 20,
  perPageRange: 's'
};
PageSizeSelector.defaultProps = defaultProps$1;
var PageSizeSelector$1 = PageSizeSelector;

var _ref = {
  name: "15sp3y3",
  styles: "flex-grow:2"
} ;
const Pagination = props => {
  const totalPages = Math.ceil(props.totalItems / props.perPage);
  const pageItems = props.page === totalPages ? props.totalItems - props.perPage * (props.page - 1) : props.perPage;
  return jsxRuntime.jsxs(Spacings__default["default"].Inline, {
    justifyContent: "space-between",
    children: [jsxRuntime.jsx("div", {
      css: _ref,
      children: jsxRuntime.jsx(PageSizeSelector$1, {
        pageItems: pageItems,
        perPage: props.perPage,
        perPageRange: props.perPageRange,
        onPerPageChange: props.onPerPageChange
      })
    }), jsxRuntime.jsx(PageNavigator$1, {
      totalPages: totalPages,
      page: props.page,
      onPageChange: props.onPageChange
    })]
  });
};
Pagination.displayName = 'Pagination';
const defaultProps = {
  perPage: 20,
  perPageRange: 's'
};
Pagination.defaultProps = defaultProps;
var Pagination$1 = Pagination;

// NOTE: This string will be replaced on build time with the package version.
var version = "16.11.0";

exports.PageNavigator = PageNavigator$1;
exports.PageSizeSelector = PageSizeSelector$1;
exports.Pagination = Pagination$1;
exports.version = version;
