// @commercetools-test-data/channel/dist/commercetools-test-data-channel.esm.js
import { Generator, fake, Transformer, Builder, sequence } from '@commercetools-test-data/core';
import { LocalizedString, Address, ClientLogging } from '@commercetools-test-data/commons';
import _slicedToArray from '@babel/runtime-corejs3/helpers/esm/slicedToArray';
import { createRelatedDates } from '@commercetools-test-data/utils';

var roles = {
  InventorySupply: 'InventorySupply',
  ProductDistribution: 'ProductDistribution',
  OrderExport: 'OrderExport',
  OrderImport: 'OrderImport',
  Primary: 'Primary'
};

var constants = /*#__PURE__*/Object.freeze({
  __proto__: null,
  roles: roles
});

// https://docs.commercetools.com/api/projects/channels#channeldraft
var generator$1 = Generator({
  fields: {
    key: fake(function (f) {
      return f.lorem.slug(2);
    }),
    roles: [roles.Primary],
    name: fake(function () {
      return LocalizedString.random();
    }),
    description: fake(function () {
      return LocalizedString.random();
    }),
    address: fake(function () {
      return Address.random();
    }),
    custom: null,
    geoLocation: null
  }
});

var transformers$1 = {
  default: Transformer('default', {
    buildFields: ['name', 'description', 'address']
  }),
  rest: Transformer('rest', {
    buildFields: ['name', 'description', 'address']
  }),
  graphql: Transformer('graphql', {
    buildFields: ['name', 'description', 'address'],
    addFields: function addFields() {
      return {
        __typename: 'ChannelDraft'
      };
    }
  })
};

var ChannelDraft = function ChannelDraft() {
  return Builder({
    generator: generator$1,
    transformers: transformers$1
  });
};

var hub = function hub() {
  return ChannelDraft().key('hub').description(LocalizedString.presets.empty()['en-US']('Hub')).name(LocalizedString.presets.empty()['en-US']('Hub')).roles([roles.ProductDistribution, roles.InventorySupply]);
};

var storeOther = function storeOther() {
  return ChannelDraft().key('store_other').description(LocalizedString.presets.empty()['en-US']('Store Other')).name(LocalizedString.presets.empty()['en-US']('Store Other')).roles([roles.ProductDistribution]);
};

var storeUsa = function storeUsa() {
  return ChannelDraft().key('store_usa').description(LocalizedString.presets.empty()['en-US']('USA Store')).name(LocalizedString.presets.empty()['en-US']('USA Store')).roles([roles.ProductDistribution]);
};

var presets$2 = {
  storeUsa: storeUsa,
  storeOther: storeOther,
  hub: hub
};

var withInventorySupplyAndProductDistributionRoles = function withInventorySupplyAndProductDistributionRoles() {
  return ChannelDraft().roles([roles.InventorySupply, roles.ProductDistribution]);
};

var withInventorySupplyAndProductDistributionRolesNoAddress = function withInventorySupplyAndProductDistributionRolesNoAddress() {
  return ChannelDraft().address(undefined).roles([roles.InventorySupply, roles.ProductDistribution]);
};

var withInventorySupplyRole = function withInventorySupplyRole() {
  return ChannelDraft().roles([roles.InventorySupply]);
};

var withInventorySupplyRoleNoAddress = function withInventorySupplyRoleNoAddress() {
  return ChannelDraft().address(undefined).roles([roles.InventorySupply]);
};

var withProductDistributionRole = function withProductDistributionRole() {
  return ChannelDraft().roles([roles.ProductDistribution]);
};

var presets$1 = {
  sampleDataFashion: presets$2,
  withProductDistributionRole: withProductDistributionRole,
  withInventorySupplyRole: withInventorySupplyRole,
  withInventorySupplyRoleNoAddress: withInventorySupplyRoleNoAddress,
  withInventorySupplyAndProductDistributionRoles: withInventorySupplyAndProductDistributionRoles,
  withInventorySupplyAndProductDistributionRolesNoAddress: withInventorySupplyAndProductDistributionRolesNoAddress
};

var index = /*#__PURE__*/Object.freeze({
  __proto__: null,
  constants: constants,
  random: ChannelDraft,
  presets: presets$1
});

var _createRelatedDates = createRelatedDates(),
  _createRelatedDates2 = _slicedToArray(_createRelatedDates, 2),
  getOlderDate = _createRelatedDates2[0],
  getNewerDate = _createRelatedDates2[1];

// https://docs.commercetools.com/api/projects/channels#channel
var generator = Generator({
  fields: {
    id: fake(function (f) {
      return f.datatype.uuid();
    }),
    key: fake(function (f) {
      return f.lorem.slug(2);
    }),
    version: sequence(),
    createdAt: fake(getOlderDate),
    lastModifiedAt: fake(getNewerDate),
    createdBy: fake(function () {
      return ClientLogging.random();
    }),
    lastModifiedBy: fake(function () {
      return ClientLogging.random();
    }),
    roles: [roles.Primary],
    name: null,
    description: null,
    address: fake(function () {
      return Address.random();
    }),
    reviewRatingStatistics: null,
    custom: null,
    geoLocation: null
  }
});

var transformers = {
  default: Transformer('default', {
    buildFields: ['createdBy', 'lastModifiedBy', 'name', 'description', 'address']
  }),
  rest: Transformer('rest', {
    buildFields: ['createdBy', 'lastModifiedBy', 'name', 'description', 'address']
  }),
  graphql: Transformer('graphql', {
    buildFields: ['address', 'createdBy', 'lastModifiedBy'],
    addFields: function addFields(_ref) {
      var fields = _ref.fields;
      var nameAllLocales = LocalizedString.toLocalizedField(fields.name);
      var descriptionAllLocales = LocalizedString.toLocalizedField(fields.description);
      return {
        __typename: 'Channel',
        nameAllLocales: nameAllLocales,
        descriptionAllLocales: descriptionAllLocales
      };
    }
  })
};

var Model = function Model() {
  return Builder({
    generator: generator,
    transformers: transformers
  });
};

var clothesStore = function clothesStore() {
  return Model().name(LocalizedString.presets.empty().en('Clothes Store')).key('clothes-store-key');
};

var foodStore = function foodStore() {
  return Model().name(LocalizedString.presets.empty().en('Food Store').de('Lebensmittelgeschäft')).key('food-store-key');
};

var presets = {
  foodStore: foodStore,
  clothesStore: clothesStore
};

export { index as ChannelDraft, constants, presets$1 as draftPresets, presets, Model as random };
