// @commercetools-uikit/text/dist/commercetools-uikit-text.cjs.prod.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _Object$keys = require('@babel/runtime-corejs3/core-js-stable/object/keys');
var _Object$getOwnPropertySymbols = require('@babel/runtime-corejs3/core-js-stable/object/get-own-property-symbols');
var _filterInstanceProperty = require('@babel/runtime-corejs3/core-js-stable/instance/filter');
var _Object$getOwnPropertyDescriptor = require('@babel/runtime-corejs3/core-js-stable/object/get-own-property-descriptor');
var _forEachInstanceProperty = require('@babel/runtime-corejs3/core-js-stable/instance/for-each');
var _Object$getOwnPropertyDescriptors = require('@babel/runtime-corejs3/core-js-stable/object/get-own-property-descriptors');
var _Object$defineProperties = require('@babel/runtime-corejs3/core-js-stable/object/define-properties');
var _Object$defineProperty = require('@babel/runtime-corejs3/core-js-stable/object/define-property');
var _defineProperty = require('@babel/runtime-corejs3/helpers/defineProperty');
require('prop-types');
var react$1 = require('react');
var reactIntl = require('react-intl');
var utils = require('@commercetools-uikit/utils');
var react = require('@emotion/react');
var designSystem = require('@commercetools-uikit/design-system');
var jsxRuntime = require('@emotion/react/jsx-runtime');

function _interopDefault (e) { return e && e.__esModule ? e : { 'default': e }; }

var _Object$keys__default = /*#__PURE__*/_interopDefault(_Object$keys);
var _Object$getOwnPropertySymbols__default = /*#__PURE__*/_interopDefault(_Object$getOwnPropertySymbols);
var _filterInstanceProperty__default = /*#__PURE__*/_interopDefault(_filterInstanceProperty);
var _Object$getOwnPropertyDescriptor__default = /*#__PURE__*/_interopDefault(_Object$getOwnPropertyDescriptor);
var _forEachInstanceProperty__default = /*#__PURE__*/_interopDefault(_forEachInstanceProperty);
var _Object$getOwnPropertyDescriptors__default = /*#__PURE__*/_interopDefault(_Object$getOwnPropertyDescriptors);
var _Object$defineProperties__default = /*#__PURE__*/_interopDefault(_Object$defineProperties);
var _Object$defineProperty__default = /*#__PURE__*/_interopDefault(_Object$defineProperty);

const getBaseStyles = () => {
  return "\n   color: ".concat(designSystem.designTokens.colorSolid, ";\n");
};
const truncate = "\n  white-space: nowrap;\n  overflow: hidden;\n  text-overflow: ellipsis;\n";
const nowrap = "\n  white-space: nowrap;\n";
const bold = "\n  font-weight: ".concat(designSystem.designTokens.fontWeightForTextAsBold, ";\n");
const italic = "\n  font-style: italic;\n";
const strikethrough = "\n  text-decoration: line-through;\n";
const getTone = tone => {
  switch (tone) {
    case 'information':
      return "color: ".concat(designSystem.designTokens.colorInfo, ";");
    case 'secondary':
      return "color: ".concat(designSystem.designTokens.colorNeutral40, ";");
    case 'positive':
      return "color: ".concat(designSystem.designTokens.colorPrimary25, ";");
    case 'primary':
      return "color: ".concat(designSystem.designTokens.colorPrimary30, ";");
    case 'negative':
      return "color: ".concat(designSystem.designTokens.colorError, ";");
    case 'inverted':
      return "color: ".concat(designSystem.designTokens.colorSurface, ";");
    case 'warning':
      return "color: ".concat(designSystem.designTokens.colorWarning, ";");
    case 'critical':
      return "color: ".concat(designSystem.designTokens.colorError40, ";");
    case 'tertiary':
      return "color: ".concat(designSystem.designTokens.colorNeutral60, ";");
    case 'inherit':
      return 'color: inherit;';
    default:
      return "";
  }
};
const getFontWeight = props => {
  switch (props.fontWeight) {
    case 'regular':
      return "font-weight: ".concat(designSystem.designTokens.fontWeight400, ";");
    case 'medium':
      return "font-weight: ".concat(designSystem.designTokens.fontWeight500, ";");
    case 'bold':
      return "font-weight: ".concat(designSystem.designTokens.fontWeight600, ";");
    default:
      if (props.isBold) {
        return "font-weight: ".concat(designSystem.designTokens.fontWeight600, ";");
      }
      return "";
  }
};
const getElementFontSize = elementType => {
  switch (elementType) {
    case 'h1':
      return designSystem.designTokens.fontSizeForTextAsH1;
    case 'h2':
      return designSystem.designTokens.fontSizeForTextAsH2;
    case 'h3':
      return designSystem.designTokens.fontSizeForTextAsH3;
    case 'h4':
      return designSystem.designTokens.fontSizeForTextAsH4;
    case 'h5':
      return designSystem.designTokens.fontSizeForTextAsH5;
    default:
      return '1rem';
  }
};
const getElementLineHeight = elementType => {
  switch (elementType) {
    case 'h1':
      return designSystem.designTokens.lineHeightForTextAsH1;
    case 'h2':
      return designSystem.designTokens.lineHeightForTextAsH2;
    case 'h3':
      return designSystem.designTokens.lineHeightForTextAsH3;
    case 'h4':
      return designSystem.designTokens.lineHeightForTextAsH4;
    case 'h5':
      return designSystem.designTokens.lineHeightForTextAsH5;
    default:
      return designSystem.designTokens.lineHeight40;
  }
};
const getElementFontWeight = elementType => {
  switch (elementType) {
    case 'h1':
      return designSystem.designTokens.fontWeightForTextAsH1;
    case 'h2':
      return designSystem.designTokens.fontWeightForTextAsH2;
    case 'h3':
      return designSystem.designTokens.fontWeightForTextAsH3;
    case 'h4':
      return designSystem.designTokens.fontWeightForTextAsH4;
    case 'h5':
      return designSystem.designTokens.fontWeightForTextAsH5;
    default:
      return 'normal';
  }
};
const bodyStyles = props => /*#__PURE__*/react.css(getBaseStyles(), " margin:0;font-size:", designSystem.designTokens.fontSizeForTextAsBody, ";line-height:", designSystem.designTokens.lineHeightForTextAsBody, ";font-weight:", designSystem.designTokens.fontWeightForTextAsBody, ";", props.isItalic && italic, " ", props.isStrikethrough && strikethrough, " ", props.tone && getTone(props.tone), " ", getFontWeight(props), " ", props.truncate && truncate, " ", props.nowrap && nowrap, ";" + ("" ), "" );
const headlineStyles = props => /*#__PURE__*/react.css(getBaseStyles(), " margin:0;font-size:", getElementFontSize(props.as), ";line-height:", getElementLineHeight(props.as), ";font-weight:", getElementFontWeight(props.as), ";", props.truncate && truncate, " ", props.nowrap && nowrap, ";" + ("" ), "" );
const subheadlineStyles = props => /*#__PURE__*/react.css(getBaseStyles(), " margin:0;font-size:", getElementFontSize(props.as), ";line-height:", getElementLineHeight(props.as), ";font-weight:", getElementFontWeight(props.as), ";", props.truncate && truncate, " ", props.nowrap && nowrap, " ", props.isBold && bold, " ", props.tone && getTone(props.tone), ";" + ("" ), "" );
const wrapStyles = () => /*#__PURE__*/react.css(getBaseStyles(), " font-size:1rem;white-space:pre-wrap;" + ("" ), "" );
const detailStyles = props => /*#__PURE__*/react.css(getBaseStyles(), " font-size:", designSystem.designTokens.fontSizeForTextAsDetail, ";line-height:", designSystem.designTokens.lineHeightForTextAsDetail, ";font-weight:", designSystem.designTokens.fontWeightForTextAsDetail, ";", props.isItalic && italic, " ", props.isStrikethrough && strikethrough, " ", props.tone && getTone(props.tone), " ", getFontWeight(props), " ", props.truncate && truncate, " ", props.nowrap && nowrap, ";" + ("" ), "" );
const captionStyles = props => /*#__PURE__*/react.css(getBaseStyles(), " font-size:", designSystem.designTokens.fontSizeForTextAsCaption, ";line-height:", designSystem.designTokens.lineHeightForTextAsCaption, ";font-weight:", designSystem.designTokens.fontWeightForTextAsCaption, ";", props.isItalic && italic, " ", props.isStrikethrough && strikethrough, " ", props.tone && getTone(props.tone), " ", getFontWeight(props), " ", props.truncate && truncate, " ", props.nowrap && nowrap, ";" + ("" ), "" );

function ownKeys(object, enumerableOnly) { var keys = _Object$keys__default["default"](object); if (_Object$getOwnPropertySymbols__default["default"]) { var symbols = _Object$getOwnPropertySymbols__default["default"](object); enumerableOnly && (symbols = _filterInstanceProperty__default["default"](symbols).call(symbols, function (sym) { return _Object$getOwnPropertyDescriptor__default["default"](object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }
function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var _context, _context2; var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? _forEachInstanceProperty__default["default"](_context = ownKeys(Object(source), !0)).call(_context, function (key) { _defineProperty(target, key, source[key]); }) : _Object$getOwnPropertyDescriptors__default["default"] ? _Object$defineProperties__default["default"](target, _Object$getOwnPropertyDescriptors__default["default"](source)) : _forEachInstanceProperty__default["default"](_context2 = ownKeys(Object(source))).call(_context2, function (key) { _Object$defineProperty__default["default"](target, key, _Object$getOwnPropertyDescriptor__default["default"](source, key)); }); } return target; }
const warnIfMissingTitle = (props, componentName) => {
  if (typeof props.title === 'string') ;
};
const warnIfMissingContent = (props, componentName) => {
  props.intlMessage || Boolean(react$1.Children.count(props.children));
};
const Text = props => jsxRuntime.jsx(jsxRuntime.Fragment, {
  children: props.intlMessage ? jsxRuntime.jsx(reactIntl.FormattedMessage, _objectSpread({}, props.intlMessage)) : props.children
});
Text.propTypes = {};
Text.displayName = 'Text';
const Headline = props => {
  warnIfMissingTitle(props);
  warnIfMissingContent(props);
  const HeadlineElement = props.as;
  if (!HeadlineElement) {
    return jsxRuntime.jsx("span", {
      id: props.id,
      children: jsxRuntime.jsx(Text, {
        intlMessage: props.intlMessage,
        children: props.children
      })
    });
  }
  return jsxRuntime.jsx(HeadlineElement, _objectSpread(_objectSpread({
    id: props.id,
    css: headlineStyles(props),
    title: props.title
  }, utils.filterDataAttributes(props)), {}, {
    children: jsxRuntime.jsx(Text, {
      intlMessage: props.intlMessage,
      children: props.children
    })
  }));
};
Headline.propTypes = {};
Headline.displayName = 'TextHeadline';
const Subheadline = props => {
  warnIfMissingTitle(props);
  warnIfMissingContent(props);
  const SubheadlineElement = props.as;
  if (!SubheadlineElement) {
    return jsxRuntime.jsx(Text, {
      id: props.id,
      intlMessage: props.intlMessage,
      children: props.children
    });
  }
  return jsxRuntime.jsx(SubheadlineElement, _objectSpread(_objectSpread({
    id: props.id,
    title: props.title,
    css: subheadlineStyles(props)
  }, utils.filterDataAttributes(props)), {}, {
    children: jsxRuntime.jsx(Text, {
      intlMessage: props.intlMessage,
      children: props.children
    })
  }));
};
Subheadline.propTypes = {};
Subheadline.displayName = 'TextSubheadline';
const Wrap = props => {
  warnIfMissingTitle(props);
  warnIfMissingContent(props);
  return jsxRuntime.jsx("div", _objectSpread(_objectSpread({
    id: props.id,
    css: wrapStyles(),
    title: props.title
  }, utils.filterDataAttributes(props)), {}, {
    children: jsxRuntime.jsx(Text, {
      intlMessage: props.intlMessage,
      children: props.children
    })
  }));
};
Wrap.propTypes = {};
Wrap.displayName = 'TextWrap';
const Body = props => {
  warnIfMissingTitle(props);
  warnIfMissingContent(props);
  if (props.as) {
    const BodyElement = props.as;
    return jsxRuntime.jsx(BodyElement, _objectSpread(_objectSpread({
      id: props.id,
      css: bodyStyles(props),
      title: props.title
    }, utils.filterDataAttributes(props)), {}, {
      children: jsxRuntime.jsx(Text, {
        intlMessage: props.intlMessage,
        children: props.children
      })
    }));
  }
  return jsxRuntime.jsx("p", _objectSpread(_objectSpread({
    id: props.id,
    css: bodyStyles(props),
    title: props.title
  }, utils.filterDataAttributes(props)), {}, {
    children: jsxRuntime.jsx(Text, {
      intlMessage: props.intlMessage,
      children: props.children
    })
  }));
};
Body.propTypes = {};
Body.displayName = 'TextBody';
const Detail = props => {
  warnIfMissingTitle(props);
  warnIfMissingContent(props);
  if (props.as) {
    const TextDetailElement = props.as;
    return jsxRuntime.jsx(TextDetailElement, _objectSpread(_objectSpread({
      id: props.id,
      css: detailStyles(props),
      title: props.title,
      "aria-labelledby": props['aria-labelledby']
    }, utils.filterDataAttributes(props)), {}, {
      children: jsxRuntime.jsx(Text, {
        intlMessage: props.intlMessage,
        children: props.children
      })
    }));
  }
  return jsxRuntime.jsx("div", _objectSpread(_objectSpread({
    css: detailStyles(props),
    title: props.title,
    "aria-labelledby": props['aria-labelledby']
  }, utils.filterDataAttributes(props)), {}, {
    children: jsxRuntime.jsx(Text, {
      intlMessage: props.intlMessage,
      children: props.children
    })
  }));
};
Detail.propTypes = {};
Detail.displayName = 'TextDetail';
const Caption = props => {
  warnIfMissingTitle(props);
  warnIfMissingContent(props);
  return jsxRuntime.jsx("div", _objectSpread(_objectSpread({
    css: captionStyles(props),
    title: props.title,
    "aria-labelledby": props['aria-labelledby']
  }, utils.filterDataAttributes(props)), {}, {
    children: jsxRuntime.jsx(Text, {
      intlMessage: props.intlMessage,
      children: props.children
    })
  }));
};
Caption.propTypes = {};
Caption.displayName = 'TextCaption';
var text = {
  Headline,
  Wrap,
  Subheadline,
  Detail,
  Body,
  Caption
};

// NOTE: This string will be replaced on build time with the package version.
var version = "16.11.0";

exports["default"] = text;
exports.version = version;
