// @commercetools-uikit/primary-button/dist/commercetools-uikit-primary-button.cjs.prod.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _Object$keys = require('@babel/runtime-corejs3/core-js-stable/object/keys');
var _Object$getOwnPropertySymbols = require('@babel/runtime-corejs3/core-js-stable/object/get-own-property-symbols');
var _filterInstanceProperty = require('@babel/runtime-corejs3/core-js-stable/instance/filter');
var _Object$getOwnPropertyDescriptor = require('@babel/runtime-corejs3/core-js-stable/object/get-own-property-descriptor');
var _forEachInstanceProperty = require('@babel/runtime-corejs3/core-js-stable/instance/for-each');
var _Object$getOwnPropertyDescriptors = require('@babel/runtime-corejs3/core-js-stable/object/get-own-property-descriptors');
var _Object$defineProperties = require('@babel/runtime-corejs3/core-js-stable/object/define-properties');
var _Object$defineProperty = require('@babel/runtime-corejs3/core-js-stable/object/define-property');
var _defineProperty = require('@babel/runtime-corejs3/helpers/defineProperty');
require('prop-types');
var react$1 = require('react');
var omit = require('lodash/omit');
var react = require('@emotion/react');
var Inline = require('@commercetools-uikit/spacings-inline');
var designSystem = require('@commercetools-uikit/design-system');
var utils = require('@commercetools-uikit/utils');
var AccessibleButton = require('@commercetools-uikit/accessible-button');
var jsxRuntime = require('@emotion/react/jsx-runtime');

function _interopDefault (e) { return e && e.__esModule ? e : { 'default': e }; }

var _Object$keys__default = /*#__PURE__*/_interopDefault(_Object$keys);
var _Object$getOwnPropertySymbols__default = /*#__PURE__*/_interopDefault(_Object$getOwnPropertySymbols);
var _filterInstanceProperty__default = /*#__PURE__*/_interopDefault(_filterInstanceProperty);
var _Object$getOwnPropertyDescriptor__default = /*#__PURE__*/_interopDefault(_Object$getOwnPropertyDescriptor);
var _forEachInstanceProperty__default = /*#__PURE__*/_interopDefault(_forEachInstanceProperty);
var _Object$getOwnPropertyDescriptors__default = /*#__PURE__*/_interopDefault(_Object$getOwnPropertyDescriptors);
var _Object$defineProperties__default = /*#__PURE__*/_interopDefault(_Object$defineProperties);
var _Object$defineProperty__default = /*#__PURE__*/_interopDefault(_Object$defineProperty);
var omit__default = /*#__PURE__*/_interopDefault(omit);
var Inline__default = /*#__PURE__*/_interopDefault(Inline);
var AccessibleButton__default = /*#__PURE__*/_interopDefault(AccessibleButton);

/* eslint-disable import/prefer-default-export */
const getSizeStyles = size => {
  switch (size) {
    // rendering the same style for now to not introduce a breaking change
    case 'small':
    case 'medium':
      return /*#__PURE__*/react.css("height:", designSystem.designTokens.heightForButtonAsMedium, ";border-radius:", designSystem.designTokens.borderRadiusForButtonAsMedium, ";padding:", designSystem.designTokens.paddingForButtonAsMedium, ";" + ("" ), "" );
    case 'big':
      return /*#__PURE__*/react.css("height:", designSystem.designTokens.heightForButtonAsBig, ";border-radius:", designSystem.designTokens.borderRadiusForButtonAsBig, ";padding:", designSystem.designTokens.paddingForButtonAsBig, ";" + ("" ), "" );
    default:
      return /*#__PURE__*/react.css("" , "" );
  }
};
const getButtonStyles = (isDisabled, isActive, tone, size) => {
  const baseStyles = /*#__PURE__*/react.css("align-items:center;color:", designSystem.designTokens.colorSurface, ";transition:background-color ", designSystem.designTokens.transitionLinear80Ms, ";", getSizeStyles(size), ";" + ("" ), "" );
  // "disabled" takes precendece over "active"
  if (isDisabled) {
    return [baseStyles, /*#__PURE__*/react.css("&,&:active,&:hover{background-color:", designSystem.designTokens.backgroundColorForButtonWhenDisabled, ";color:", designSystem.designTokens.colorNeutral60, ";box-shadow:", designSystem.designTokens.shadowForButtonWhenDisabled, ";}" + ("" ), "" )];
  }
  if (isActive) {
    const baseActiveStyles = [baseStyles, /*#__PURE__*/react.css("box-shadow:", designSystem.designTokens.shadowForButtonWhenActive, ";&:hover,&:focus{box-shadow:", designSystem.designTokens.shadowForButtonWhenFocused, ";}" + ("" ), "" )];
    switch (tone) {
      case 'primary':
        return [baseActiveStyles, /*#__PURE__*/react.css("background-color:", designSystem.designTokens.backgroundColorForButtonAsPrimaryWhenActive, ";&:focus,&:hover{background-color:", designSystem.designTokens.backgroundColorForButtonAsPrimaryWhenHovered, ";}" + ("" ), "" )];
      case 'urgent':
        return [baseActiveStyles, /*#__PURE__*/react.css("background-color:", designSystem.designTokens.backgroundColorForButtonAsUrgentWhenActive, ";&:focus,&:hover{background-color:", designSystem.designTokens.backgroundColorForButtonAsUrgentWhenHovered, ";}" + ("" ), "" )];
      case 'critical':
        return [baseActiveStyles, /*#__PURE__*/react.css("background-color:", designSystem.designTokens.backgroundColorForButtonAsCriticalWhenActive, ";&:focus,&:hover{background-color:", designSystem.designTokens.backgroundColorForButtonAsCriticalWhenHovered, ";}" + ("" ), "" )];
      default:
        return baseActiveStyles;
    }
  }
  const baseDefaultStyles = [baseStyles, /*#__PURE__*/react.css("box-shadow:", designSystem.designTokens.shadowForButton, ";&:hover,&:focus{box-shadow:", designSystem.designTokens.shadowForButtonWhenFocused, ";}&:active{box-shadow:", designSystem.designTokens.shadowForButtonWhenActive, ";}" + ("" ), "" )];
  switch (tone) {
    case 'primary':
      return [baseDefaultStyles, /*#__PURE__*/react.css("background-color:", designSystem.designTokens.colorPrimary, ";&:focus,&:hover{background-color:", designSystem.designTokens.backgroundColorForButtonAsPrimaryWhenHovered, ";}&:active{background-color:", designSystem.designTokens.backgroundColorForButtonAsPrimaryWhenActive, ";}" + ("" ), "" )];
    case 'urgent':
      return [baseDefaultStyles, /*#__PURE__*/react.css("background-color:", designSystem.designTokens.colorWarning, ";&:focus,&:hover{background-color:", designSystem.designTokens.backgroundColorForButtonAsUrgentWhenHovered, ";}&:active{background-color:", designSystem.designTokens.backgroundColorForButtonAsUrgentWhenActive, ";}" + ("" ), "" )];
    case 'critical':
      return [baseDefaultStyles, /*#__PURE__*/react.css("background-color:", designSystem.designTokens.colorError, ";&:focus,&:hover{background-color:", designSystem.designTokens.backgroundColorForButtonAsCriticalWhenHovered, ";}&:active{background-color:", designSystem.designTokens.backgroundColorForButtonAsCriticalWhenActive, ";}" + ("" ), "" )];
    default:
      return baseDefaultStyles;
  }
};

function ownKeys(object, enumerableOnly) { var keys = _Object$keys__default["default"](object); if (_Object$getOwnPropertySymbols__default["default"]) { var symbols = _Object$getOwnPropertySymbols__default["default"](object); enumerableOnly && (symbols = _filterInstanceProperty__default["default"](symbols).call(symbols, function (sym) { return _Object$getOwnPropertyDescriptor__default["default"](object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }
function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var _context, _context2; var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? _forEachInstanceProperty__default["default"](_context = ownKeys(Object(source), !0)).call(_context, function (key) { _defineProperty(target, key, source[key]); }) : _Object$getOwnPropertyDescriptors__default["default"] ? _Object$defineProperties__default["default"](target, _Object$getOwnPropertyDescriptors__default["default"](source)) : _forEachInstanceProperty__default["default"](_context2 = ownKeys(Object(source))).call(_context2, function (key) { _Object$defineProperty__default["default"](target, key, _Object$getOwnPropertyDescriptor__default["default"](source, key)); }); } return target; }
const propsToOmit = ['type'];
const defaultProps = {
  type: 'button',
  size: 'big',
  isToggleButton: false,
  tone: 'primary'
};
const PrimaryButton = props => {
  const buttonAttributes = _objectSpread(_objectSpread({
    'data-track-component': 'PrimaryButton'
  }, utils.filterInvalidAttributes(omit__default["default"](props, propsToOmit))), {}, {
    // if there is a divergence between `isDisabled` and `disabled`,
    // we fall back to `isDisabled`
    disabled: props.isDisabled
  });
  utils.useWarning(!Boolean(props.size === 'small'), 'PrimaryButton `small` value for `size` property has been renamed to `medium`. Please update that value when using this component');
  const isActive = Boolean(props.isToggleButton && props.isToggled);
  return jsxRuntime.jsx(AccessibleButton__default["default"], {
    as: props.as,
    type: props.type,
    buttonAttributes: buttonAttributes,
    label: props.label,
    onClick: props.onClick,
    isToggleButton: props.isToggleButton,
    isToggled: props.isToggled,
    isDisabled: props.isDisabled,
    css: getButtonStyles(props.isDisabled, isActive, props.tone, props.size),
    children: jsxRuntime.jsxs(Inline__default["default"], {
      alignItems: "center",
      scale: "xs",
      children: [Boolean(props.iconLeft) && jsxRuntime.jsx("span", {
        css: /*#__PURE__*/react.css("margin:0 ", designSystem.designTokens.spacing10, " 0 0;display:flex;align-items:center;justify-content:center;" + ("" ), "" ),
        children: props.iconLeft && /*#__PURE__*/react$1.cloneElement(props.iconLeft, {
          color: props.isDisabled ? 'neutral60' : 'surface',
          size: props.size === 'big' ? 'big' : 'medium'
        })
      }), jsxRuntime.jsx("span", {
        children: props.label
      })]
    })
  });
};
PrimaryButton.propTypes = {};
PrimaryButton.defaultProps = defaultProps;
PrimaryButton.displayName = 'PrimaryButton';
var PrimaryButton$1 = PrimaryButton;

// NOTE: This string will be replaced on build time with the package version.
var version = "16.11.0";

exports["default"] = PrimaryButton$1;
exports.version = version;
